---
title: "Getting started with HoneyBADGER"
author: "Jean Fan"
date: '`r Sys.Date()`'
output:
  pdf_document: default
  html_document: default
md_document:
  variant: markdown_github
vignette: |
  %\VignetteIndexEntry{Vignette Title} %\VignetteEngine{knitr::rmarkdown} \usepackage[utf8]{inputenc}
---


```{r, include = FALSE}
library(knitr)
opts_chunk$set(
    warning = FALSE,
    message = FALSE,
    fig.path = 'figure/',
    cache.path = 'cache/',
    cache = TRUE,
    dpi=100
)
```


```{r}
library(HoneyBADGER)
```

As an example, we will load prepared data for tumor cells from patient MGH31 from [Patel et al](http://science.sciencemag.org/content/344/6190/1396).

First, we will load the gene expression matrices for tumor cells along with a normal expression reference derived from averaging normal brain samples found in [GTex](https://www.gtexportal.org/home/). We also load a corresponding biomaRt instance (for human) to obtain chromosomal coordinate information for our genes. 

```{r}
data(gexp) ## tumor cells
data(ref) ## reference

require(biomaRt) ## for gene coordinates
mart.obj <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = 'hsapiens_gene_ensembl', host = "jul2015.archive.ensembl.org")

print(gexp[1:5,1:5])
print(ref[1:5])
print(mart.obj)
```

Make a new `HoneyBADGER` object and initialize the gene expression matrices. The data has already been filtered for highly expressed shared genes and scaled for library size differences so we can override the default filtering and scaling.

```{r, fig.width=12, fig.height=4}
hb <- new('HoneyBADGER', name='MGH31')
hb$setGexpMats(gexp, ref, mart.obj, filter=FALSE, scale=FALSE, verbose=TRUE)
hb$plotGexpProfile() ## initial visualization
```

Visually we may already visually suspect some chromosomal abnormalities on Chr 7, 10, 13, 14. We can model the gene expression variance and use a HMM to identify regions affected by CNVs.

```{r}
hb$setMvFit(verbose=TRUE) ## model variance
hb$setGexpDev(verbose=TRUE) ## model necessary expression deviation to identify CNVs
hb$calcGexpCnvBoundaries(init=TRUE, verbose=TRUE) ## HMM

## double check what CNVs were identified
bgf <- get('bound.genes.final', slot(hb, '.xData'))
genes <- get('genes', slot(hb, '.xData'))
regions.genes <- range(genes[unlist(bgf)])
print(regions.genes)
```

Indeed, our initial HMM has identified a number of candidate CNVs to test. We can now retest all identified CNVs on all cells to derive the final posterior probability of each CNV in each cell and plot.

```{r}
hb$retestIdentifiedCnvs()

## look at final results
results <- hb$summarizeResults(geneBased=TRUE, alleleBased=FALSE)

rgs <- range(hb$genes[unlist(hb$bound.genes.final),])
      names <- apply(as.data.frame(rgs), 1, paste0, collapse=":")
      retest <- hb$results[['gene-based']]
      amp.gexp.prob <- do.call(rbind, lapply(retest, function(x) x[[1]]))
      del.gexp.prob <- do.call(rbind, lapply(retest, function(x) x[[2]]))

      ## filter to regions with at least some highly confident cells
      vi1 <- rowSums(amp.gexp.prob > 0.75) > min.num.cells
      amp.gexp.prob <- amp.gexp.prob[vi1,] ## amplifications
      vi2 <- rowSums(del.gexp.prob > 0.75) > min.num.cells
      del.gexp.prob <- del.gexp.prob[vi2,] ## amplifications
      
      rownames(amp.gexp.prob) <- paste0('amp', names[vi1])
      rownames(del.gexp.prob) <- paste0('del', names[vi2])

      results <- rbind(amp.gexp.prob, del.gexp.prob)      
hc <- hclust(dist(colSums(results)))
order <- hc$labels[hc$order]
```


```{r, fig.width=12, fig.height=4}
## plot all chromosomes
hb$plotGexpProfile(cellOrder=order)
```

```{r, fig.width=12, fig.height=4}
## plot identified CNV region only
hb$plotGexpProfile(cellOrder=order, region=regions.genes[vi1,]) ## amplifications
hb$plotGexpProfile(cellOrder=order, region=regions.genes[vi2,]) ## deletions
```


```{r, fig.width=12, fig.height=8}
## visualize as heatmap 
hb$visualizeResults(geneBased=TRUE, alleleBased=FALSE, hc=hc, margins=c(25,15))
```

We can also analyze using allele information. Load the allele data and add them to our HoneyBADGER object.

```{r}
data(r)
data(cov.sc)

hb <- new('HoneyBADGER', name='MGH31')
hb$setAlleleMats(r.init=r, n.sc.init=cov.sc, n.cores=1, het.deviance.threshold=0.1)

library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
hb$setGeneFactors(txdb)
```


```{r, fig.width=12, fig.height=4}

hb$plotSmoothedAlleleProfile(cellOrder=hc$labels[hc$order])
hb$plotAlleleProfile(cellOrder=hc$labels[hc$order])

```

Now run our allele HMM to identify putative regions affected by deletions or LOH.

```{r}
hb$calcAlleleCnvBoundaries(init=TRUE)

## double check what CNVs were identified
bsf <- get('bound.snps.final', slot(hb, '.xData'))
snps <- get('snps', slot(hb, '.xData'))
regions.snp <- range(snps[unlist(bsf)])
print(regions.snp)
```

We can again retest.

```{r}
hb$retestIdentifiedCnvs(retestBoundGenes=FALSE, retestBoundSnps=TRUE)

retest <- get('results', slot(hb, '.xData'))[['allele-based']]
results <- do.call(rbind, lapply(retest, function(x) x))
```

```{r, fig.width=12, fig.height=4}
hb$plotAlleleProfile(cellOrder=hc$labels[hc$order])

hb$plotAlleleProfile(cellOrder=hc$labels[hc$order], region=regions.snp[1])
hb$plotAlleleProfile(cellOrder=hc$labels[hc$order], region=regions.snp[5])
```

```{r}
hb$summarizeResults(geneBased=FALSE, alleleBased=TRUE)
```

```{r, fig.width=12, fig.height=4}
## visualize as heatmap 
hb$visualizeResults(geneBased=FALSE, alleleBased=TRUE, hc=hc)
```
