---
title: "Getting started with HoneyBADGER"
author: "Jean Fan"
date: "2017-03-06"
md_document:
  variant: markdown_github
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---






```r
library(HoneyBADGER)
```

Load gene expression data for cells from patient MGH31 along with an expression reference.


```r
data(gexp)
data(ref)

require(biomaRt) ## for gene coordinates
mart.obj <- useMart(biomart = "ENSEMBL_MART_ENSEMBL", dataset = 'hsapiens_gene_ensembl', host = "jul2015.archive.ensembl.org")
```

Make new HoneyBADGER object and initialzie the gene expression matrices. The data has already been filtered for highly expressed shared genes and scaled for library size differences so we can override the default filtering and scaling.


```r
hb <- HoneyBADGER$new()
hb$setGexpMats(gexp, ref, mart.obj, filter=FALSE, scale=FALSE)
```

```
## Initializing expression matrices ... 
## normalizing gene expression for 6082 genes and 75 cells ... 
## Done setting initial expression matrices!
```

```r
hb$plotGexpProfile()
```

<img src="figure/unnamed-chunk-4-1.png" title="plot of chunk unnamed-chunk-4" alt="plot of chunk unnamed-chunk-4" width="800px" height="400px" />

Now we can model the gene expression variance and use a HMM to identify regions affected by CNVs.


```r
hb$setMvFit()
```

```
## Modeling expected variance ... Done!
```

```r
hb$calcGexpCnvBoundaries(init=TRUE)
```

```
## max vote:4GENES AFFECTED BY CNV: ADCY2 MTRR SEMA5A FAM173B CCT5 CMBL MARCH6 CTNND2 TRIO ANKH FAM134B MYO10 BASP1 GUSBP1 CDH10 CDH6 DROSHA C5orf22 PDZD2 MTMR12 ZFR SUB1 TARS C1QTNF3 RAI14 RAD1 BRIX1 LMBRD2 SKP2 NIPBL C5orf42 NUP155 WDR70 LIFR OSMR RICTOR TTC33 PRKAA1 RPL37 PLCXD3 OXCT1 C5orf51 SEPP1 ZNF131 HMGCS1 PAIP1 NNT HCN1 PARP8 ITGA2 MOCS2 NDUFS4 DHX29 SKIV2L2 PPAP2A IL6ST MAP3K1 GPBP1 PDE4D ELOVL7 SMIM15 ZSWIM6 DIMT1 KIF2A IPO11 RNF180 SREK1IP1 CWC27 PPWD1 TRIM23 NLN ERBB2IP SREK1 MAST4 PIK3R1 SLC30A5 RAD17 OCLN NAIP MAP1B MRPS27 TNPO1 BTF3 ANKRA2 GFM2 NSA2 FAM169A HMGCR COL4A3BP POLK PDE8B WDR41 TBCA AP3B1 SCAMP1 ARSB JMY HOMER1 PAPD4 CMYA5 THBS4 ZFYVE16 MTRNR2L2 MSH3 RASGRF2 SSBP2 RPS23 TMEM167A VCAN COX7C RASA1 CCNH TMEM161B LINC00461 MEF2C CETN3 LYSMD3 ARRDC3 FAM172A TTC37 RHOBTB3 GLRX ELL2 PCSK1 CAST LNPEP LIX1 RIOK2 CHD1 PAM PPIP5K2 EFNA5 FER PJA2 MAN2A1 WDR36 STARD4 NREP APC REEP5 DCP2 MCC YTHDC2 TRIM36 FEM1C CDO1 ATG12 AP3S1 COMMD10 SEMA6A DMXL1 HSD17B4 SNCAIP SNX2 SNX24 CSNK1G3 ZNF608 GRAMD3 ALDH7A1 PHAX PRRC1 SLC12A2 HINT1 LYRM7 CDC42SE2 FNIP1 ACSL6 P4HA2 IRF1 RAD50 KIF3A UQCRQ AFF4 ZCCHC10 HSPA4 C5orf15 VDAC1 SKP1 UBE2B CDKN2AIPNL SAR1B SEC24A CAMLG DDX46 TXNDC15 H2AFY CXCL14 SMAD5 SPOCK1 KLHL3 FAM13B BRD8 CDC23 KDM3B EGR1 ETF1 HSPA9 CTNNA1 SIL1 MATR3 PAIP2 DNAJC18 PURA CYSTM1 PFDN1 SRA1 SLC35A4 IK HARS HARS2 ZMAT2 TAF7 PCDHGC3 DIAPH1 HDAC3 KIAA0141 RNF14 GNPDA1 NDFIP1 SPRY4 FGF1 ARHGAP26 NR3C1 YIPF5 LARS RBM27 TCERG1 PPP2R2B STK32A DPYSL3 JAKMIP2 FBXO38 SH3TC2 GRPEL2 CSNK1A1 SLC26A2 CAMK2A TCOF1 RPS14 NDST1 RBM22 DCTN4 GPX3 CCDC69 GM2A SLC36A1 ATOX1 G3BP1 GRIA1 FAM114A2 MFAP3 LARP1 CNOT8 GEMIN5 MRPL22 SGCD HAVCR2 CYFIP2 ADAM19 CLINT1 EBF1 RNF145 UBLCP1 TTC1 PWWP2A SLU7 ATP10B GABRB2 GABRG2 CCNG1 NUDCD2 MAT2B TENM2 WWC1 RARS SLIT3 KCNIP1 NPM1 STK10 DUSP1 ERGIC1 RPL26L1 ATP6V0E1 CREBRF CPEB4 SFXN1 KIAA1191 HIGD2A FAF2 UIMC1 ZNF346 NSD1 LMAN2 DBN1 DDX41 TMED9 FAM153C NHP2 CLK4 RUFY1 HNRNPH1 CANX SQSTM1 GFPT2 CNOT6 LINC00847 BTNL9 TRIM7 GNB2L1 EXOC2 SERPINB1 SERPINB9 SERPINB6 NQO2 RIPK1 BPHL TUBB2A TUBB2B PRPF4B ECI2 CDYL LYRM4 NRN1 RIOK1 SNRNP48 PAK1IP1 TMEM14C TMEM14B ELOVL2 NEDD9 TMEM170B SIRT5 RANBP9 CD83 ATXN1 CAP2 NUP153aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0.80675 0.97375 0.5455 0.624 0.90075 0.30225 0.421 0.75125 0.98375 0.9925 0.95 0.676 0.155 0.99875 0.152 0.99425 0.879 0.974 0.63875 0.87325 0.7765 0.62425 0.6705 0.9835 0.97725 0.8265 0.94025 0.68025 0.6775 0.96325 0.57125 0.9665 0.8315 0.75825 0.56575 0.72325 0.19875 0.67125 0.6135 0.568 0.89575 0.33275 0.84075 0.61 0.65625 0.7215 0.6885 0.34025 0.75825 0.978 0.75075 0.96625 0.14 0.7515 0.98325 0.77275 0.7555 0.4465 0.77525 0.96475 0.35775 0.7455 0.96275 0.99625 0.9885 0.96 0.733 0.87625 0.9835 0.578 0.999 0.46425 0.87775 0.88625 0.96325DELETION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0GENES AFFECTED BY CNV: COX19 ZFAND2A PSMG3 FTSJ2 EIF3B GNA12 FOXK1 AP5Z1 MMD2 RNF216P1 FBXL18 RNF216 USP42 RAC1 KDELR2 ZDHHC4 C1GALT1 GLCCI1 ICA1 NDUFA4 PHF14 TMEM106B ARL4A ETV1 ANKMY2 BZW2 SNX13 HDAC9 TWISTNB SP4 RAPGEF5 TOMM7 FAM126A KLHL7 NUPL2 GPNMB TRA2A FAM221A DFNA5 OSBPL3 CYCS HNRNPA2B1 CBX3 HIBADH TAX1BP1 CREB5 CPVL WIPF3 SCRN1 FKBP14 GGCT GARS ADCYAP1R1 PDE1C LSM5 AVL9 KBTBD2 RP9P FKBP9 BBS9 BMPER DPY19L1 HERPUD2 SEPT7 ANLN ELMO1 STARD3NL VPS41 YAE1D1 RALA LINC00265 CDK13 MRPL32 STK17A COA1 BLVRA URGCP UBE2D4 POLR2J4 DBNL POLD2 NUDCD3 DDX56 TMED4 OGDH PPIA H2AFV PURB SNHG15 TBRG4 RAMP3 ADCY1 SEPT7P2 IGFBP3 HUS1 UPP1 FIGNL1 GRB10 SEC61G EGFR LANCL2 VOPP1 GBAS PSPH CCT6A SUMF2 CHCHD2 ZNF680 ZNF138 ZNF273 GUSB TPST1 LINC00174 KCTD7 TMEM248 SBDS TYW1 CALN1 SBDSP1 POM121 BAZ1B TBL2 WBSCR22 ABHD11 ELN EIF4H LAT2 HIP1 RHBDD2 POR TMEM120A STYXL1 MDH2 HSPB1 YWHAG FGL2 PTPN12 PHTF2 MAGI2 GNAI1 CACNA2D1 PCLO GRM3 KIAA1324L DMTF1 CROT SLC25A40 ADAM22 SRI STEAP2 CLDN12 CDK14 AKAP9 KRIT1 ANKIB1 GATAD1 PEX1 RBM48 CDK6 SAMD9L CCDC132 GNG11 BET1 COL1A2 CASD1 SGCE PEG10 PPP1R9A PON2 SLC25A13 SHFM1 ASNS LMTK2 TECPR1 TMEM130 TRRAP ARPC1A ARPC1B PDAP1 BUD31 CPSF4 ATP5J2 ZNF394 ZKSCAN5 FAM200A ZNF655 TRIM4 ZKSCAN1 ZNF3 COPS6 MCM7 TAF6 CNPY4 LAMTOR4 PILRB TSC22D4 AGFG2 GIGYF1 POP7 SRRT PLOD3 ZNHIT1 FIS1 CUX1 POLR2J ARMC10 NAPEPLD PMPCB DNAJC2 PSMC2 RELN ORC5 LHFPL3 SRPK2 PUS7 RINT1 ATXN7L1 CDHR3 SYPL1 NAMPT HBP1 COG5 BCAP29 CBLL1 DLD LAMB1 NRCAM PNPLA8 THAP5 DNAJB9 LRRN3 DOCK4 ZNF277 IFRD1 TMEM168 MDFIC CAV2 CAV1 CAPZA2 ST7 CTTNBP2 ING3 FAM3C AASS CADPS2 NDUFA5 WASL POT1 ARF5 SND1 RBM28 HILPDA METTL2B ATP6V1F TNPO3 AHCYL2 STRIP2 UBE2H ZC3HC1 KLHDC10 TMEM209 CEP41 MKLN1 PODXL PLXNA4 CHCHD3 EXOC4 SLC35B4 AKR1B1 BPGM CALD1 WDR91 CNOT4 NUP205 C7orf73 PTN CREB3L2 TRIM24 ZC3HAV1 UBN2 HIPK2 SLC37A3 MKRN1 DENND2A NDUFB2 BRAF MRPS33 AGK GSTK1 CASP2 ZYX CNTNAP2 CUL1 PDIA4 ZNF783 ATP6V0E2 TMEM176B CDK5 ABCF2 NUB1 RHEB PRKAG2 GALNT11 DPP6 NOM1 UBE3C DNAJB6 PTPRN2 NCAPG2 ESYT2 WDR60aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0.9785 0.9575 0.28475 0.99825 0.992 0.99575 0.31075 0.98325 0.9865 0.99125 0.957 0.9925 0.22675 0.998 0.55525 0.99 0.9885 0.9765 0.9735 0.993 0.99775 0.99275 0.99825 0.98725 0.921 0.916 0.987 0.6925 0.99775 0.986 0.97975 0.96825 0.98175 0.982 0.98 0.9745 0.98675 0.98925 0.28775 0.9925 0.9955 0.4385 0.945 0.96975 0.677 0.98375 0.99775 0.9925 0.985 0.988 0.96175 0.98525 0.14925 0.993 0.9595 0.9915 0.83075 0.8185 0.9 0.989 0.99875 0.996 0.9095 0.86925 0.98925 0.991 0.9675 0.979 0.98625 0.455 0.9575 0.98725 0.97475 0.99625 0.9885DELETION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0max vote:4GENES AFFECTED BY CNV: CACNA1B DIP2C LARP4B GTPBP4 IDI1 ADARB2 PFKP PITRM1 KLF6 GDI2 FBXO18 RBM17 PFKFB3 SFMBT2 ITIH5 ATP5C1 CELF2 UPF2 DHTKD1 NUDT5 CDC123 CAMK1D OPTN PHYH NMT2 PTER RSU1 VIM CACNB2 MLLT10 PIP4K2A MSRB2 KIAA1217 ARHGAP21 PRTFDC1 THNSL1 ABI1 ANKRD26 YME1L1 ACBD5 RAB18 MPP7 WAC KIAA1462 MTPAP ZEB1 ARHGAP12 KIF5B ITGB1 NRP1 CUL2 CREM ZNF33A ZNF37A BMS1 HNRNPF C10orf10 RASSF4 ZNF22 TIMM23 NCOA4 SYT15 CISD1 TFAM PHYHIPL SLC16A9 ANK3 ARID5B ZNF365 NRBF2 JMJD1C LRRTM3 HERC4 HNRNPH3 RUFY2 CCAR1 DDX50 DDX21 HK1 TYSND1 SAR1A PPA1 SGPL1 UNC5B CDH23 PSAP CHST3 SPOCK2 ASCC1 ANAPC16 DDIT4 MICU1 MCU P4HA1 ECD MRPS16 USP54 SEC24C CAMK2G VCL AP3M1 ADK KAT6B SAMD8 VDAC2 ZNF503 KCNMA1 POLR3A RPS24 ZMIZ1 PPIF EIF5AL1 TMEM254 ANXA11 FAM213A TSPAN14 GHITM CCSER2 MMRN2 GLUD1 ATAD1 PTEN ACTA2 LIPA IFIT3 IFIT1 PANK1 RPP30 PCGF5 HECTD2 TNKS2 IDE MYOF FRA10AC1 PLCE1 SORBS1 ALDH18A1 TCTN3 ENTPD1 TM9SF3 LCOR PGAM1 MMS19 MORN4 GOLGA7B R3HCC1L CNNM1 COX15 DNMBP CWF19L1 BLOC1S2 SCD NDUFB8 HIF1AN FAM178A MRPL43 PDZD7 SFXN3 BTRC DPCD MGEA5 C10orf76 NOLC1 CUEDC2 TMEM180 ACTR1A ARL3 WBP1L NT5C2 USMG5 SLK GSTO1 XPNPEP1 ADD3 SMC3 PDCD4 SHOC2 ZDHHC6 TCF7L2 CASP7 ABLIM1 FAM160B1 EMX2 FAM204A CACUL1 NANOS1 EIF3A PRDX3 TIAL1 INPP5F SEC23IP WDR11 FGFR2 TACC2 PLEKHA1 ACADSB BUB3 OAT CTBP2 BCCIP DOCK1 PTPRE GLRX3 BNIP3 LRRC27 TUBGCP2 ECHS1 BET1L RIC8A SIRT3 PSMD13 IFITM2 IFITM3 TALDO1 PDDC1 SLC25A22 RPLP2 CD151 POLR2L CHID1 CTSD CD81aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0DELETION PROBABILITY: 0.98275 0.989 0.46375 0.99075 0.98875 0.97725 0.59625 0.99175 0.981 0.9965 0.97775 0.99775 0.587 0.99625 0.2275 0.979 0.99375 0.9895 0.94 0.988 0.99075 0.97975 0.999 0.9865 0.98625 0.87 0.99475 0.98175 0.99225 0.98725 0.97575 0.997 0.992 1 0.9775 0.99475 0.893 0.97925 0.29275 0.9995 0.955 0.65025 0.79925 0.995 0.35725 0.98825 0.9985 0.9705 0.96125 0.9955 0.99225 0.9955 0.17625 0.9805 0.9965 0.99875 0.84775 0.24175 0.9985 0.99925 0.995 0.9975 0.98525 0.9775 0.99725 0.99675 0.9255 0.82 0.99725 0.57125 0.9975 0.99175 0.978 0.998 0.996GENES AFFECTED BY CNV: CRYL1 IFT88 XPO4 SAP18 SACS SPATA13 PARP4 MTMR6 NUPL1 RNF6 WASF3 GPR12 MTIF3 PAN3 FLT1 POMP KATNAL1 USPL1 HSPH1 FRY N4BP2L2 NBEA DCLK1 SPG20 SMAD9 ALG5 EXOSC8 SUPT20H UFM1 NHLRC3 LHFP COG6 FOXO1 MRPS31 KBTBD6 NAA16 AKAP11 DNAJC15 TSC22D1 GTF2F2 TPT1 ZC3H13 ESD SUCLA2 NUDT15 MED4 ITM2B LPAR6 RCBTB2 FNDC3A SETDB2 PHF11 RCBTB1 KPNA3 SPRYD7 FAM124A CKAP2 HNRNPA1L2 SUGT1 PCDH17 PCDH9 DIS3 PIBF1 KLF12 COMMD6 MYCBP2 EDNRB RBM26 NDFIP2 SPRY2 SLITRK1 GPR180 CLDN10 DZIP1 DNAJC3 UGGT2 MBNL2 IPO5 UBAC2 TM9SF2 TPP2 ARGLU1 CARKD ANKRD10 ARHGEF7 ATP11A PCID2 DCUN1D2 TMCO3 UPF3A TTC5 CCNB1IP1 PARP2 TEP1 OSGEP APEX1 METTL17 NDRG2 LINC00641 HNRNPC SUPT16H CHD8 RAB2B TOX4 METTL3 DAD1 ABHD4 OXA1L MRPL52 RBM23 PRMT5 PSMB5 ACIN1 C14orf119 SLC7A8 NGDN AP1G2 DHRS4L2 DCAF11 PSME1 PSME2 CHMP4A GMPR2 TINF2 RABGGTA DHRS1 SDR39U1 NOVA1 FOXG1 PRKD1 G2E3 SCFD1 STRN3 AP4S1 HECTD1 HEATR5A DTD2 NUBPL ARHGAP5 AKAP6 NPAS3 EGLN3 SPTSSA EAPP SNX6 CFL2 BAZ1A SRP54 FAM177A1 PPP2R3C NFKBIA RALGAPA1 BRMS1L MBIP SEC23A TRAPPC6B PNN KLHL28 FAM179B PRPF39 FKBP3 RPS29 RPL36AL NEMF ARF6 MAP4K5 ATL1 NIN PYGL TRIM9 TMX1 C14orf166 TXNDC16 ERO1L PSMC6 STYX GNPNAT1 FERMT2 DDHD1 GMFB CGRRF1 SAMD4A SOCS4 MAPK1IP1L LGALS3 ATG14 KTN1 EXOC5 AP5M1 ACTR10 PSMA3 ARID4A KIAA0586 DAAM1 JKAMP PCNXL4 DHRS7 PPM1A MNAT1 TRMT5 SLC38A6 HIF1A RHOJ SYNE2 MTHFD1 HSPA2 SPTB MAX FUT8 GPHN MPP5 ATP6V1D EIF2S1 PLEKHH1 VTI1B RDH11 ZFP36L1 ACTN1 DCAF5 EXD2 ERH SLC39A9 SRSF5 COX16 MED6 MAP3K9 PCNX SIPA1L1 RGS6 DPF3 ZFYVE1 RBM25 PSEN1 PAPLN NUMB ELMSAN1 FAM161B ALDH6A1 LIN52 ABCD4 NPC2 ISCA2 FCF1 RPS6KL1 EIF2B2 MLH3 NEK9 TMED10 C14orf1 TTLL5 TGFB3 IFT43 GPATCH2L ANGEL1 VIPAS39 AHSA1 SPTLC2 SLIRP SNW1 NRXN3 DIO2 STON2 SEL1L GALC KCNK10 ZC3H14 TTC8 FOXN3 TDP1 PSMC1 CALM1 C14orf159 SMEK1 FBLN5 TRIP11 ATXN3 NDUFB1 CPSF2 LGMN GOLGA5 UNC79 SERPINA3 CLMN C14orf132 GSKIP PAPOLA SETD3 YY1 WARS WDR25 MEG3 PPP2R5C DYNC1H1 HSP90AA1 WDR20 MOK CINP TECPR2 TNFAIP2 EIF5 MARK3 CKB BAG5 APOPT1 KLC1 ZFYVE21 C14orf2aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0DELETION PROBABILITY: 0.93175 0.614 0.36725 0.999 0.99975 0.9995 0.41025 0.9975 0.80425 0.72775 0.9735 0.99525 0.352 0.86075 0.25175 0.908 1 0.92625 0.9865 0.48475 0.84025 0.9965 0.99775 0.813 0.9005 0.99125 0.70175 0.994 0.90125 0.8285 0.8295 0.67525 0.996 0.6145 0.98625 0.99325 0.9905 0.996 0.563 0.9975 0.8805 0.1835 0.93425 0.78375 0.84525 0.999 0.99925 0.9935 0.99475 0.44975 0.99925 0.451 0.40025 0.999 0.74 0.99075 0.5415 0.1885 0.989 0.63525 0.99975 0.9995 0.6605 0.91225 0.62525 0.881 0.993 0.99575 0.8765 0.32575 0.41825 0.998 0.992 0.997 0.7675[1] "CNV SNPS:"
##   [1] "CACNA1B"  "DIP2C"    "LARP4B"   "GTPBP4"   "IDI1"     "ADARB2"  
##   [7] "PFKP"     "PITRM1"   "KLF6"     "GDI2"     "FBXO18"   "RBM17"   
##  [13] "PFKFB3"   "SFMBT2"   "ITIH5"    "ATP5C1"   "CELF2"    "UPF2"    
##  [19] "DHTKD1"   "NUDT5"    "CDC123"   "CAMK1D"   "OPTN"     "PHYH"    
##  [25] "NMT2"     "PTER"     "RSU1"     "VIM"      "CACNB2"   "MLLT10"  
##  [31] "PIP4K2A"  "MSRB2"    "KIAA1217" "ARHGAP21" "PRTFDC1"  "THNSL1"  
##  [37] "ABI1"     "ANKRD26"  "YME1L1"   "ACBD5"    "RAB18"    "MPP7"    
##  [43] "WAC"      "KIAA1462" "MTPAP"    "ZEB1"     "ARHGAP12" "KIF5B"   
##  [49] "ITGB1"    "NRP1"     "CUL2"     "CREM"     "ZNF33A"   "ZNF37A"  
##  [55] "BMS1"     "HNRNPF"   "C10orf10" "RASSF4"   "ZNF22"    "TIMM23"  
##  [61] "NCOA4"    "SYT15"    "CISD1"    "TFAM"     "PHYHIPL"  "SLC16A9" 
##  [67] "ANK3"     "ARID5B"   "ZNF365"   "NRBF2"    "JMJD1C"   "LRRTM3"  
##  [73] "HERC4"    "HNRNPH3"  "RUFY2"    "CCAR1"    "DDX50"    "DDX21"   
##  [79] "HK1"      "TYSND1"   "SAR1A"    "PPA1"     "SGPL1"    "UNC5B"   
##  [85] "CDH23"    "PSAP"     "CHST3"    "SPOCK2"   "ASCC1"    "ANAPC16" 
##  [91] "DDIT4"    "MICU1"    "MCU"      "P4HA1"    "ECD"      "MRPS16"  
##  [97] "USP54"    "SEC24C"   "CAMK2G"   "VCL"      "AP3M1"    "ADK"     
## [103] "KAT6B"    "SAMD8"    "VDAC2"    "ZNF503"   "KCNMA1"   "POLR3A"  
## [109] "RPS24"    "ZMIZ1"    "PPIF"     "EIF5AL1"  "TMEM254"  "ANXA11"  
## [115] "FAM213A"  "TSPAN14"  "GHITM"    "CCSER2"   "MMRN2"    "GLUD1"   
## [121] "ATAD1"    "PTEN"     "ACTA2"    "LIPA"     "IFIT3"    "IFIT1"   
## [127] "PANK1"    "RPP30"    "PCGF5"    "HECTD2"   "TNKS2"    "IDE"     
## [133] "MYOF"     "FRA10AC1" "PLCE1"    "SORBS1"   "ALDH18A1" "TCTN3"   
## [139] "ENTPD1"   "TM9SF3"   "LCOR"     "PGAM1"    "MMS19"    "MORN4"   
## [145] "GOLGA7B"  "R3HCC1L"  "CNNM1"    "COX15"    "DNMBP"    "CWF19L1" 
## [151] "BLOC1S2"  "SCD"      "NDUFB8"   "HIF1AN"   "FAM178A"  "MRPL43"  
## [157] "PDZD7"    "SFXN3"    "BTRC"     "DPCD"     "MGEA5"    "C10orf76"
## [163] "NOLC1"    "CUEDC2"   "TMEM180"  "ACTR1A"   "ARL3"     "WBP1L"   
## [169] "NT5C2"    "USMG5"    "SLK"      "GSTO1"    "XPNPEP1"  "ADD3"    
## [175] "SMC3"     "PDCD4"    "SHOC2"    "ZDHHC6"   "TCF7L2"   "CASP7"   
## [181] "ABLIM1"   "FAM160B1" "EMX2"     "FAM204A"  "CACUL1"   "NANOS1"  
## [187] "EIF3A"    "PRDX3"    "TIAL1"    "INPP5F"   "SEC23IP"  "WDR11"   
## [193] "FGFR2"    "TACC2"    "PLEKHA1"  "ACADSB"   "BUB3"     "OAT"     
## [199] "CTBP2"    "BCCIP"    "DOCK1"    "PTPRE"    "GLRX3"    "BNIP3"   
## [205] "LRRC27"   "TUBGCP2"  "ECHS1"    "BET1L"    "RIC8A"    "SIRT3"   
## [211] "PSMD13"   "IFITM2"   "IFITM3"   "TALDO1"   "PDDC1"    "SLC25A22"
## [217] "RPLP2"    "CD151"    "POLR2L"   "CHID1"    "CTSD"     "CD81"    
##      MGH31_A02 MGH31_A03 MGH31_A04 MGH31_A05 MGH31_A06 MGH31_A07 MGH31_A08
## [1,]   0.98275     0.989   0.46375   0.99075   0.98875   0.97725   0.59625
##      MGH31_A10 MGH31_A11 MGH31_B01 MGH31_B02 MGH31_B04 MGH31_B05 MGH31_B06
## [1,]   0.99175     0.981    0.9965   0.97775   0.99775     0.587   0.99625
##      MGH31_B07 MGH31_B08 MGH31_B09 MGH31_B10 MGH31_C02 MGH31_C03 MGH31_C04
## [1,]    0.2275     0.979   0.99375    0.9895      0.94     0.988   0.99075
##      MGH31_C05 MGH31_C06 MGH31_C07 MGH31_C09 MGH31_C10 MGH31_C11 MGH31_D02
## [1,]   0.97975     0.999    0.9865   0.98625      0.87   0.99475   0.98175
##      MGH31_D03 MGH31_D05 MGH31_D07 MGH31_D09 MGH31_D10 MGH31_D11 MGH31_D12
## [1,]   0.99225   0.98725   0.97575     0.997     0.992         1    0.9775
##      MGH31_E01 MGH31_E02 MGH31_E03 MGH31_E04 MGH31_E05 MGH31_E08 MGH31_E09
## [1,]   0.99475     0.893   0.97925   0.29275    0.9995     0.955   0.65025
##      MGH31_E11 MGH31_E12 MGH31_F01 MGH31_F02 MGH31_F03 MGH31_F04 MGH31_F06
## [1,]   0.79925     0.995   0.35725   0.98825    0.9985    0.9705   0.96125
##      MGH31_F07 MGH31_F08 MGH31_F09 MGH31_F10 MGH31_F11 MGH31_F12 MGH31_G01
## [1,]    0.9955   0.99225    0.9955   0.17625    0.9805    0.9965   0.99875
##      MGH31_G02 MGH31_G04 MGH31_G05 MGH31_G07 MGH31_G08 MGH31_G09 MGH31_G10
## [1,]   0.84775   0.24175    0.9985   0.99925     0.995    0.9975   0.98525
##      MGH31_G12 MGH31_H01 MGH31_H02 MGH31_H04 MGH31_H05 MGH31_H06 MGH31_H07
## [1,]    0.9775   0.99725   0.99675    0.9255      0.82   0.99725   0.57125
##      MGH31_H08 MGH31_H09 MGH31_H10 MGH31_H11 MGH31_H12
## [1,]    0.9975   0.99175     0.978     0.998     0.996
## [1] "GROUP1:"
##  [1] "MGH31_A02" "MGH31_A03" "MGH31_A05" "MGH31_A06" "MGH31_A07"
##  [6] "MGH31_A10" "MGH31_A11" "MGH31_B01" "MGH31_B02" "MGH31_B04"
## [11] "MGH31_B06" "MGH31_B08" "MGH31_B09" "MGH31_B10" "MGH31_C02"
## [16] "MGH31_C03" "MGH31_C04" "MGH31_C05" "MGH31_C06" "MGH31_C07"
## [21] "MGH31_C09" "MGH31_C10" "MGH31_C11" "MGH31_D02" "MGH31_D03"
## [26] "MGH31_D05" "MGH31_D07" "MGH31_D09" "MGH31_D10" "MGH31_D11"
## [31] "MGH31_D12" "MGH31_E01" "MGH31_E02" "MGH31_E03" "MGH31_E05"
## [36] "MGH31_E08" "MGH31_E11" "MGH31_E12" "MGH31_F02" "MGH31_F03"
## [41] "MGH31_F04" "MGH31_F06" "MGH31_F07" "MGH31_F08" "MGH31_F09"
## [46] "MGH31_F11" "MGH31_F12" "MGH31_G01" "MGH31_G02" "MGH31_G05"
## [51] "MGH31_G07" "MGH31_G08" "MGH31_G09" "MGH31_G10" "MGH31_G12"
## [56] "MGH31_H01" "MGH31_H02" "MGH31_H04" "MGH31_H05" "MGH31_H06"
## [61] "MGH31_H08" "MGH31_H09" "MGH31_H10" "MGH31_H11" "MGH31_H12"
## [1] "GROUP2:"
## [1] "MGH31_B07" "MGH31_F10" "MGH31_G04"
## [1] "Recursion for Group1"
## max vote:6GENES AFFECTED BY CNV: ADCY2 MTRR SEMA5A FAM173B CCT5 CMBL MARCH6 CTNND2 TRIO ANKH FAM134B MYO10 BASP1 GUSBP1 CDH10 CDH6 DROSHA C5orf22 PDZD2 MTMR12 ZFR SUB1 TARS C1QTNF3 RAI14 RAD1 BRIX1 LMBRD2 SKP2 NIPBL C5orf42 NUP155 WDR70 LIFR OSMR RICTOR TTC33 PRKAA1 RPL37 PLCXD3 OXCT1 C5orf51 SEPP1 ZNF131 HMGCS1 PAIP1 NNT HCN1 PARP8 ITGA2 MOCS2 NDUFS4 DHX29 SKIV2L2 PPAP2A IL6ST MAP3K1 GPBP1 PDE4D ELOVL7 SMIM15 ZSWIM6 DIMT1 KIF2A IPO11 RNF180 SREK1IP1 CWC27 PPWD1 TRIM23 NLN ERBB2IP SREK1 MAST4 PIK3R1 SLC30A5 RAD17 OCLN NAIP MAP1B MRPS27 TNPO1 BTF3 ANKRA2 GFM2 NSA2 FAM169A HMGCR COL4A3BP POLK PDE8B WDR41 TBCA AP3B1 SCAMP1 ARSB JMY HOMER1 PAPD4 CMYA5 THBS4 ZFYVE16 MTRNR2L2 MSH3 RASGRF2 SSBP2 RPS23 TMEM167A VCAN COX7C RASA1 CCNH TMEM161B LINC00461 MEF2C CETN3 LYSMD3 ARRDC3 FAM172A TTC37 RHOBTB3 GLRX ELL2 PCSK1 CAST LNPEP LIX1 RIOK2 CHD1 PAM PPIP5K2 EFNA5 FER PJA2 MAN2A1 WDR36 STARD4 NREP APC REEP5 DCP2 MCC YTHDC2 TRIM36 FEM1C CDO1 ATG12 AP3S1 COMMD10 SEMA6A DMXL1 HSD17B4 SNCAIP SNX2 SNX24 CSNK1G3 ZNF608 GRAMD3 ALDH7A1 PHAX PRRC1 SLC12A2 HINT1 LYRM7 CDC42SE2 FNIP1 ACSL6 P4HA2 IRF1 RAD50 KIF3A UQCRQ AFF4 ZCCHC10 HSPA4 C5orf15 VDAC1 SKP1 UBE2B CDKN2AIPNL SAR1B SEC24A CAMLG DDX46 TXNDC15 H2AFY CXCL14 SMAD5 SPOCK1 KLHL3 FAM13B BRD8 CDC23 KDM3B EGR1 ETF1 HSPA9 CTNNA1 SIL1 MATR3 PAIP2 DNAJC18 PURA CYSTM1 PFDN1 SRA1 SLC35A4 IK HARS HARS2 ZMAT2 TAF7 PCDHGC3 DIAPH1 HDAC3 KIAA0141 RNF14 GNPDA1 NDFIP1 SPRY4 FGF1 ARHGAP26 NR3C1 YIPF5 LARS RBM27 TCERG1 PPP2R2B STK32A DPYSL3 JAKMIP2 FBXO38 SH3TC2 GRPEL2 CSNK1A1 SLC26A2 CAMK2A TCOF1 RPS14 NDST1 RBM22 DCTN4 GPX3 CCDC69 GM2A SLC36A1 ATOX1 G3BP1 GRIA1 FAM114A2 MFAP3 LARP1 CNOT8 GEMIN5 MRPL22 SGCD HAVCR2 CYFIP2 ADAM19 CLINT1 EBF1 RNF145 UBLCP1 TTC1 PWWP2A SLU7 ATP10B GABRB2 GABRG2 CCNG1 NUDCD2 MAT2B TENM2 WWC1 RARS SLIT3 KCNIP1 NPM1 STK10 DUSP1 ERGIC1 RPL26L1 ATP6V0E1 CREBRF CPEB4 SFXN1 KIAA1191 HIGD2A FAF2 UIMC1 ZNF346 NSD1 LMAN2 DBN1 DDX41 TMED9 FAM153C NHP2 CLK4 RUFY1 HNRNPH1 CANX SQSTM1 GFPT2 CNOT6 LINC00847 BTNL9 TRIM7 GNB2L1 EXOC2 SERPINB1 SERPINB9 SERPINB6 NQO2 RIPK1 BPHL TUBB2A TUBB2B PRPF4B ECI2 CDYL LYRM4 NRN1 RIOK1 SNRNP48 PAK1IP1 TMEM14C TMEM14B ELOVL2 NEDD9 TMEM170B SIRT5 RANBP9 CD83 ATXN1 CAP2 NUP153aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0.808 0.972 0.60575 0.90575 0.291 0.75525 0.98225 0.99225 0.94375 0.6775 0.997 0.993 0.86975 0.97525 0.62475 0.8755 0.775 0.6015 0.668 0.98725 0.984 0.8295 0.93125 0.68275 0.656 0.96225 0.58875 0.962 0.82825 0.75225 0.5845 0.722 0.214 0.6895 0.55375 0.89225 0.844 0.62125 0.73475 0.71575 0.3655 0.7655 0.97525 0.74225 0.96975 0.75025 0.98425 0.77175 0.777 0.77325 0.9635 0.36275 0.7415 0.959 0.99775 0.99075 0.95825 0.742 0.8865 0.9855 0.9975 0.4905 0.8865 0.88 0.96975DELETION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0GENES AFFECTED BY CNV: SUN1 COX19 ZFAND2A PSMG3 FTSJ2 EIF3B GNA12 FOXK1 AP5Z1 MMD2 RNF216P1 FBXL18 RNF216 USP42 RAC1 KDELR2 ZDHHC4 C1GALT1 GLCCI1 ICA1 NDUFA4 PHF14 TMEM106B ARL4A ETV1 ANKMY2 BZW2 SNX13 HDAC9 TWISTNB SP4 RAPGEF5 TOMM7 FAM126A KLHL7 NUPL2 GPNMB TRA2A FAM221A DFNA5 OSBPL3 CYCS HNRNPA2B1 CBX3 HIBADH TAX1BP1 CREB5 CPVL WIPF3 SCRN1 FKBP14 GGCT GARS ADCYAP1R1 PDE1C LSM5 AVL9 KBTBD2 RP9P FKBP9 BBS9 BMPER DPY19L1 HERPUD2 SEPT7 ANLN ELMO1 STARD3NL VPS41 YAE1D1 RALA LINC00265 CDK13 MRPL32 STK17A COA1 BLVRA URGCP UBE2D4 POLR2J4 DBNL POLD2 NUDCD3 DDX56 TMED4 OGDH PPIA H2AFV PURB SNHG15 TBRG4 RAMP3 ADCY1 SEPT7P2 IGFBP3 HUS1 UPP1 FIGNL1 GRB10 SEC61G EGFR LANCL2 VOPP1 GBAS PSPH CCT6A SUMF2 CHCHD2 ZNF680 ZNF138 ZNF273 GUSB TPST1 LINC00174 KCTD7 TMEM248 SBDS TYW1 CALN1 SBDSP1 POM121 BAZ1B TBL2 WBSCR22 ABHD11 ELN EIF4H LAT2 HIP1 RHBDD2 POR TMEM120A STYXL1 MDH2 HSPB1 YWHAG FGL2 PTPN12 PHTF2 MAGI2 GNAI1 CACNA2D1 PCLO GRM3 KIAA1324L DMTF1 CROT SLC25A40 ADAM22 SRI STEAP2 CLDN12 CDK14 AKAP9 KRIT1 ANKIB1 GATAD1 PEX1 RBM48 CDK6 SAMD9L CCDC132 GNG11 BET1 COL1A2 CASD1 SGCE PEG10 PPP1R9A PON2 SLC25A13 SHFM1 ASNS LMTK2 TECPR1 TMEM130 TRRAP ARPC1A ARPC1B PDAP1 BUD31 CPSF4 ATP5J2 ZNF394 ZKSCAN5 FAM200A ZNF655 TRIM4 ZKSCAN1 ZNF3 COPS6 MCM7 TAF6 CNPY4 LAMTOR4 PILRB TSC22D4 AGFG2 GIGYF1 POP7 SRRT PLOD3 ZNHIT1 FIS1 CUX1 POLR2J ARMC10 NAPEPLD PMPCB DNAJC2 PSMC2 RELN ORC5 LHFPL3 SRPK2 PUS7 RINT1 ATXN7L1 CDHR3 SYPL1 NAMPT HBP1 COG5 BCAP29 CBLL1 DLD LAMB1 NRCAM PNPLA8 THAP5 DNAJB9 LRRN3 DOCK4 ZNF277 IFRD1 TMEM168 MDFIC CAV2 CAV1 CAPZA2 ST7 CTTNBP2 ING3 FAM3C AASS CADPS2 NDUFA5 WASL POT1 ARF5 SND1 RBM28 HILPDA METTL2B ATP6V1F TNPO3 AHCYL2 STRIP2 UBE2H ZC3HC1 KLHDC10 TMEM209 CEP41 MKLN1 PODXL PLXNA4 CHCHD3 EXOC4 SLC35B4 AKR1B1 BPGM CALD1 WDR91 CNOT4 NUP205 C7orf73 PTN CREB3L2 TRIM24 ZC3HAV1 UBN2 HIPK2 SLC37A3 MKRN1 DENND2A NDUFB2 BRAF MRPS33 AGK GSTK1 CASP2 ZYX CNTNAP2 CUL1 PDIA4 ZNF783 ATP6V0E2 TMEM176B CDK5 ABCF2 NUB1 RHEB PRKAG2 GALNT11 DPP6 NOM1 UBE3C DNAJB6 PTPRN2 NCAPG2 ESYT2 WDR60aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0.98 0.955 0.9985 0.99625 0.99525 0.98375 0.98575 0.9925 0.966 0.99325 0.99675 0.98975 0.98825 0.98125 0.97875 0.99525 0.99775 0.9935 0.99775 0.9905 0.927 0.917 0.9885 0.709 0.99625 0.99 0.98125 0.9625 0.98175 0.9825 0.97775 0.97175 0.9895 0.9895 0.993 0.99725 0.955 0.9645 0.989 0.998 0.995 0.98725 0.97925 0.96775 0.987 0.99575 0.964 0.99275 0.82875 0.89875 0.99075 0.9975 0.9975 0.92275 0.8835 0.98725 0.99175 0.9695 0.98525 0.98625 0.96 0.987 0.97975 0.9975 0.9865DELETION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0GENES AFFECTED BY CNV: ZNF211 ZNF671 ZNF814 ZNF418 ZNF606 ZNF135 ZNF329 ZNF544 ZNF497 RPS5 CHMP2A SOX12 NRSN2 TBC1D20 CSNK2A1 PSMF1 FKBP1A NSFL1C STK35 SNRPB ZNF343 NOP56 IDH3B VPS16 PTPRA UBOX5 ITPA C20orf194 ATRN ADAM33 AP5S1 MAVS RNF24 SMOX PRNP RASSF2 SLC23A2 TMEM230 PCNA CDS2 GPCPD1 CRLS1 PLCB4 MKKS JAG1 BTBD3 ESF1 FLRT3 SNRPB2 PCSK2 DSTN SNX5 ZNF133 DZANK1 POLR3F RBBP9 SEC23B LINC00493 DTD1 RIN2 NAA20 CRNKL1 RALGAPA2 XRN2 CST3 APMAP ENTPD6 PYGB ZNF337 FRG1B HM13 BCL2L1 TM9SF4 POFUT1 KIF3B ASXL1 MAPRE1 CDK5RAP1 CBFA2T2 EIF2S2 AHCY ITCH DYNLRB1 PIGU NCOA6 GSS MYH7B EDEM2 EIF6 CEP250 ERGIC3 CPNE1 NFS1 ROMO1 RBM39 PHF20 LINC00657 EPB41L1 AAR2 DLGAP4 NDRG3 SOGA1 SAMHD1 RPN2 MANBAL BLCAP CTNNBL1 TTI1 KIAA1755 RALGAPB TOP1 ZHX3 CHD6 PTPRT L3MBTL1 IFT52 GDAP1L1 TTPAL SERINC3 PKIG YWHAB TOMM34 SDC4 SYS1 PIGT ACOT8 ZSWIM1 CTSA PLTP PCIF1 SLC35C2 ELMO2 ZNF334 TP53RK ZMYND8 SULF2 ARFGEF2 CSE1L STAU1 DDX27 ZNFX1 ZFAS1 B4GALT5 SLC9A8 RNF114 TMEM189 BCAS4 DPM1 ATP9A ZFP64 BCAS1 PFDN4 DOK5 FAM210B CSTF1 RTFDC1 BMP7 RAE1 RAB22A VAPB STX16 GNAS NELFCD ATP5E SLMO2 FAM217B PSMA7 SS18L1 OSBPL2 ADRM1 RPS21 GID8 NKAIN4 HELZ2 STMN3 TPD52L2 PRPF6 PCMTD2 CBS SMIM11 HSPA13 NRIP1aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0.97075 0.72475 0.94925 0.95825 0.9965 0.9895 0.665 0.4615 0.98925 0.9805 0.8915 0.7055 0.98775 0.6465 0.91375 0.941 0.94225 0.987 0.95125 0.34925 0.472 0.882 0.82525 0.825 0.7815 0.3585 0.9795 0.757 0.97475 0.983 0.99175 0.88825 0.97575 0.982 0.996 0.9915 0.90625 0.97375 0.987 0.9955 0.85825 0.96425 0.8085 0.96825 0.94375 0.9395 0.771 0.97225 0.79025 0.97075 0.79725 0.986 0.995 0.61175 0.56075 0.8175 0.83475 0.9695 0.82025 0.62775 0.7195 0.99 0.95175 0.99975 0.74325DELETION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0max vote:2GENES AFFECTED BY CNV: CRYL1 IFT88 XPO4 SAP18 SACS SPATA13 PARP4 MTMR6 NUPL1 RNF6 WASF3 GPR12 MTIF3 PAN3 FLT1 POMP KATNAL1 USPL1 HSPH1 FRY N4BP2L2 NBEA DCLK1 SPG20 SMAD9 ALG5 EXOSC8 SUPT20H UFM1 NHLRC3 LHFP COG6 FOXO1 MRPS31 KBTBD6 NAA16 AKAP11 DNAJC15 TSC22D1 GTF2F2 TPT1 ZC3H13 ESD SUCLA2 NUDT15 MED4 ITM2B LPAR6 RCBTB2 FNDC3A SETDB2 PHF11 RCBTB1 KPNA3 SPRYD7 FAM124A CKAP2 HNRNPA1L2 SUGT1 PCDH17 PCDH9 DIS3 PIBF1 KLF12 COMMD6 MYCBP2 EDNRB RBM26 NDFIP2 SPRY2 SLITRK1 GPR180 CLDN10 DZIP1 DNAJC3 UGGT2 MBNL2 IPO5 UBAC2 TM9SF2 TPP2 ARGLU1 CARKD ANKRD10 ARHGEF7 ATP11A PCID2 DCUN1D2 TMCO3 UPF3A TTC5 CCNB1IP1 PARP2 TEP1 OSGEP APEX1 METTL17 NDRG2 LINC00641 HNRNPC SUPT16H CHD8 RAB2B TOX4 METTL3 DAD1 ABHD4 OXA1L MRPL52 RBM23 PRMT5 PSMB5 ACIN1 C14orf119 SLC7A8 NGDN AP1G2 DHRS4L2 DCAF11 PSME1 PSME2 CHMP4A GMPR2 TINF2 RABGGTA DHRS1 SDR39U1 NOVA1 FOXG1 PRKD1 G2E3 SCFD1 STRN3 AP4S1 HECTD1 HEATR5A DTD2 NUBPL ARHGAP5 AKAP6 NPAS3 EGLN3 SPTSSA EAPP SNX6 CFL2 BAZ1A SRP54 FAM177A1 PPP2R3C NFKBIA RALGAPA1 BRMS1L MBIP SEC23A TRAPPC6B PNN KLHL28 FAM179B PRPF39 FKBP3 RPS29 RPL36AL NEMF ARF6 MAP4K5 ATL1 NIN PYGL TRIM9 TMX1 C14orf166 TXNDC16 ERO1L PSMC6 STYX GNPNAT1 FERMT2 DDHD1 GMFB CGRRF1 SAMD4A SOCS4 MAPK1IP1L LGALS3 ATG14 KTN1 EXOC5 AP5M1 ACTR10 PSMA3 ARID4A KIAA0586 DAAM1 JKAMP PCNXL4 DHRS7 PPM1A MNAT1 TRMT5 SLC38A6 HIF1A RHOJ SYNE2 MTHFD1 HSPA2 SPTB MAX FUT8 GPHN MPP5 ATP6V1D EIF2S1 PLEKHH1 VTI1B RDH11 ZFP36L1 ACTN1 DCAF5 EXD2 ERH SLC39A9 SRSF5 COX16 MED6 MAP3K9 PCNX SIPA1L1 RGS6 DPF3 ZFYVE1 RBM25 PSEN1 PAPLN NUMB ELMSAN1 FAM161B ALDH6A1 LIN52 ABCD4 NPC2 ISCA2 FCF1 RPS6KL1 EIF2B2 MLH3 NEK9 TMED10 C14orf1 TTLL5 TGFB3 IFT43 GPATCH2L ANGEL1 VIPAS39 AHSA1 SPTLC2 SLIRP SNW1 NRXN3 DIO2 STON2 SEL1L GALC KCNK10 ZC3H14 TTC8 FOXN3 TDP1 PSMC1 CALM1 C14orf159 SMEK1 FBLN5 TRIP11 ATXN3 NDUFB1 CPSF2 LGMN GOLGA5 UNC79 SERPINA3 CLMN C14orf132 GSKIP PAPOLA SETD3 YY1 WARS WDR25 MEG3 PPP2R5C DYNC1H1 HSP90AA1 WDR20 MOK CINP TECPR2 TNFAIP2 EIF5 MARK3 CKB BAG5 APOPT1 KLC1 ZFYVE21 C14orf2 INF2aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0DELETION PROBABILITY: 0.92275 0.6115 0.9995 0.99975 0.99925 0.99575 0.81075 0.7165 0.9725 0.9965 0.86325 0.905 1 0.91275 0.988 0.4715 0.836 0.99625 0.9965 0.8045 0.904 0.98875 0.69475 0.9935 0.88375 0.82 0.84025 0.71825 0.99675 0.59475 0.99375 0.994 0.9905 0.99675 0.99875 0.8745 0.94025 0.78525 0.9985 0.99825 0.99425 0.99625 0.46025 0.999 0.44675 0.99925 0.729 0.9885 0.52325 0.98525 0.65425 0.99875 0.99975 0.6445 0.92075 0.64375 0.8865 0.9925 0.99225 0.8725 0.418 0.997 0.9925 0.9985 0.786[1] "CNV SNPS:"
##   [1] "SUN1"      "COX19"     "ZFAND2A"   "PSMG3"     "FTSJ2"    
##   [6] "EIF3B"     "GNA12"     "FOXK1"     "AP5Z1"     "MMD2"     
##  [11] "RNF216P1"  "FBXL18"    "RNF216"    "USP42"     "RAC1"     
##  [16] "KDELR2"    "ZDHHC4"    "C1GALT1"   "GLCCI1"    "ICA1"     
##  [21] "NDUFA4"    "PHF14"     "TMEM106B"  "ARL4A"     "ETV1"     
##  [26] "ANKMY2"    "BZW2"      "SNX13"     "HDAC9"     "TWISTNB"  
##  [31] "SP4"       "RAPGEF5"   "TOMM7"     "FAM126A"   "KLHL7"    
##  [36] "NUPL2"     "GPNMB"     "TRA2A"     "FAM221A"   "DFNA5"    
##  [41] "OSBPL3"    "CYCS"      "HNRNPA2B1" "CBX3"      "HIBADH"   
##  [46] "TAX1BP1"   "CREB5"     "CPVL"      "WIPF3"     "SCRN1"    
##  [51] "FKBP14"    "GGCT"      "GARS"      "ADCYAP1R1" "PDE1C"    
##  [56] "LSM5"      "AVL9"      "KBTBD2"    "RP9P"      "FKBP9"    
##  [61] "BBS9"      "BMPER"     "DPY19L1"   "HERPUD2"   "SEPT7"    
##  [66] "ANLN"      "ELMO1"     "STARD3NL"  "VPS41"     "YAE1D1"   
##  [71] "RALA"      "LINC00265" "CDK13"     "MRPL32"    "STK17A"   
##  [76] "COA1"      "BLVRA"     "URGCP"     "UBE2D4"    "POLR2J4"  
##  [81] "DBNL"      "POLD2"     "NUDCD3"    "DDX56"     "TMED4"    
##  [86] "OGDH"      "PPIA"      "H2AFV"     "PURB"      "SNHG15"   
##  [91] "TBRG4"     "RAMP3"     "ADCY1"     "SEPT7P2"   "IGFBP3"   
##  [96] "HUS1"      "UPP1"      "FIGNL1"    "GRB10"     "SEC61G"   
## [101] "EGFR"      "LANCL2"    "VOPP1"     "GBAS"      "PSPH"     
## [106] "CCT6A"     "SUMF2"     "CHCHD2"    "ZNF680"    "ZNF138"   
## [111] "ZNF273"    "GUSB"      "TPST1"     "LINC00174" "KCTD7"    
## [116] "TMEM248"   "SBDS"      "TYW1"      "CALN1"     "SBDSP1"   
## [121] "POM121"    "BAZ1B"     "TBL2"      "WBSCR22"   "ABHD11"   
## [126] "ELN"       "EIF4H"     "LAT2"      "HIP1"      "RHBDD2"   
## [131] "POR"       "TMEM120A"  "STYXL1"    "MDH2"      "HSPB1"    
## [136] "YWHAG"     "FGL2"      "PTPN12"    "PHTF2"     "MAGI2"    
## [141] "GNAI1"     "CACNA2D1"  "PCLO"      "GRM3"      "KIAA1324L"
## [146] "DMTF1"     "CROT"      "SLC25A40"  "ADAM22"    "SRI"      
## [151] "STEAP2"    "CLDN12"    "CDK14"     "AKAP9"     "KRIT1"    
## [156] "ANKIB1"    "GATAD1"    "PEX1"      "RBM48"     "CDK6"     
## [161] "SAMD9L"    "CCDC132"   "GNG11"     "BET1"      "COL1A2"   
## [166] "CASD1"     "SGCE"      "PEG10"     "PPP1R9A"   "PON2"     
## [171] "SLC25A13"  "SHFM1"     "ASNS"      "LMTK2"     "TECPR1"   
## [176] "TMEM130"   "TRRAP"     "ARPC1A"    "ARPC1B"    "PDAP1"    
## [181] "BUD31"     "CPSF4"     "ATP5J2"    "ZNF394"    "ZKSCAN5"  
## [186] "FAM200A"   "ZNF655"    "TRIM4"     "ZKSCAN1"   "ZNF3"     
## [191] "COPS6"     "MCM7"      "TAF6"      "CNPY4"     "LAMTOR4"  
## [196] "PILRB"     "TSC22D4"   "AGFG2"     "GIGYF1"    "POP7"     
## [201] "SRRT"      "PLOD3"     "ZNHIT1"    "FIS1"      "CUX1"     
## [206] "POLR2J"    "ARMC10"    "NAPEPLD"   "PMPCB"     "DNAJC2"   
## [211] "PSMC2"     "RELN"      "ORC5"      "LHFPL3"    "SRPK2"    
## [216] "PUS7"      "RINT1"     "ATXN7L1"   "CDHR3"     "SYPL1"    
## [221] "NAMPT"     "HBP1"      "COG5"      "BCAP29"    "CBLL1"    
## [226] "DLD"       "LAMB1"     "NRCAM"     "PNPLA8"    "THAP5"    
## [231] "DNAJB9"    "LRRN3"     "DOCK4"     "ZNF277"    "IFRD1"    
## [236] "TMEM168"   "MDFIC"     "CAV2"      "CAV1"      "CAPZA2"   
## [241] "ST7"       "CTTNBP2"   "ING3"      "FAM3C"     "AASS"     
## [246] "CADPS2"    "NDUFA5"    "WASL"      "POT1"      "ARF5"     
## [251] "SND1"      "RBM28"     "HILPDA"    "METTL2B"   "ATP6V1F"  
## [256] "TNPO3"     "AHCYL2"    "STRIP2"    "UBE2H"     "ZC3HC1"   
## [261] "KLHDC10"   "TMEM209"   "CEP41"     "MKLN1"     "PODXL"    
## [266] "PLXNA4"    "CHCHD3"    "EXOC4"     "SLC35B4"   "AKR1B1"   
## [271] "BPGM"      "CALD1"     "WDR91"     "CNOT4"     "NUP205"   
## [276] "C7orf73"   "PTN"       "CREB3L2"   "TRIM24"    "ZC3HAV1"  
## [281] "UBN2"      "HIPK2"     "SLC37A3"   "MKRN1"     "DENND2A"  
## [286] "NDUFB2"    "BRAF"      "MRPS33"    "AGK"       "GSTK1"    
## [291] "CASP2"     "ZYX"       "CNTNAP2"   "CUL1"      "PDIA4"    
## [296] "ZNF783"    "ATP6V0E2"  "TMEM176B"  "CDK5"      "ABCF2"    
## [301] "NUB1"      "RHEB"      "PRKAG2"    "GALNT11"   "DPP6"     
## [306] "NOM1"      "UBE3C"     "DNAJB6"    "PTPRN2"    "NCAPG2"   
## [311] "ESYT2"     "WDR60"    
##      MGH31_A02 MGH31_A03 MGH31_A05 MGH31_A06 MGH31_A07 MGH31_A10 MGH31_A11
## [1,]      0.98     0.955    0.9985   0.99625   0.99525   0.98375   0.98575
##      MGH31_B01 MGH31_B02 MGH31_B04 MGH31_B06 MGH31_B08 MGH31_B09 MGH31_B10
## [1,]    0.9925     0.966   0.99325   0.99675   0.98975   0.98825   0.98125
##      MGH31_C02 MGH31_C03 MGH31_C04 MGH31_C05 MGH31_C06 MGH31_C07 MGH31_C09
## [1,]   0.97875   0.99525   0.99775    0.9935   0.99775    0.9905     0.927
##      MGH31_C10 MGH31_C11 MGH31_D02 MGH31_D03 MGH31_D05 MGH31_D07 MGH31_D09
## [1,]     0.917    0.9885     0.709   0.99625      0.99   0.98125    0.9625
##      MGH31_D10 MGH31_D11 MGH31_D12 MGH31_E01 MGH31_E02 MGH31_E03 MGH31_E05
## [1,]   0.98175    0.9825   0.97775   0.97175    0.9895    0.9895     0.993
##      MGH31_E08 MGH31_E11 MGH31_E12 MGH31_F02 MGH31_F03 MGH31_F04 MGH31_F06
## [1,]   0.99725     0.955    0.9645     0.989     0.998     0.995   0.98725
##      MGH31_F07 MGH31_F08 MGH31_F09 MGH31_F11 MGH31_F12 MGH31_G01 MGH31_G02
## [1,]   0.97925   0.96775     0.987   0.99575     0.964   0.99275   0.82875
##      MGH31_G05 MGH31_G07 MGH31_G08 MGH31_G09 MGH31_G10 MGH31_G12 MGH31_H01
## [1,]   0.89875   0.99075    0.9975    0.9975   0.92275    0.8835   0.98725
##      MGH31_H02 MGH31_H04 MGH31_H05 MGH31_H06 MGH31_H08 MGH31_H09 MGH31_H10
## [1,]   0.99175    0.9695   0.98525   0.98625      0.96     0.987   0.97975
##      MGH31_H11 MGH31_H12
## [1,]    0.9975    0.9865
## [1] "GROUP1:"
##  [1] "MGH31_A02" "MGH31_A03" "MGH31_A05" "MGH31_A06" "MGH31_A07"
##  [6] "MGH31_A10" "MGH31_A11" "MGH31_B01" "MGH31_B02" "MGH31_B04"
## [11] "MGH31_B06" "MGH31_B08" "MGH31_B09" "MGH31_B10" "MGH31_C02"
## [16] "MGH31_C03" "MGH31_C04" "MGH31_C05" "MGH31_C06" "MGH31_C07"
## [21] "MGH31_C09" "MGH31_C10" "MGH31_C11" "MGH31_D03" "MGH31_D05"
## [26] "MGH31_D07" "MGH31_D09" "MGH31_D10" "MGH31_D11" "MGH31_D12"
## [31] "MGH31_E01" "MGH31_E02" "MGH31_E03" "MGH31_E05" "MGH31_E08"
## [36] "MGH31_E11" "MGH31_E12" "MGH31_F02" "MGH31_F03" "MGH31_F04"
## [41] "MGH31_F06" "MGH31_F07" "MGH31_F08" "MGH31_F09" "MGH31_F11"
## [46] "MGH31_F12" "MGH31_G01" "MGH31_G02" "MGH31_G05" "MGH31_G07"
## [51] "MGH31_G08" "MGH31_G09" "MGH31_G10" "MGH31_G12" "MGH31_H01"
## [56] "MGH31_H02" "MGH31_H04" "MGH31_H05" "MGH31_H06" "MGH31_H08"
## [61] "MGH31_H09" "MGH31_H10" "MGH31_H11" "MGH31_H12"
## [1] "GROUP2:"
## character(0)
## [1] "Recursion for Group1"
## max vote:2GENES AFFECTED BY CNV: ADCY2 MTRR SEMA5A FAM173B CCT5 CMBL MARCH6 CTNND2 TRIO ANKH FAM134B MYO10 BASP1 GUSBP1 CDH10 CDH6 DROSHA C5orf22 PDZD2 MTMR12 ZFR SUB1 TARS C1QTNF3 RAI14 RAD1 BRIX1 LMBRD2 SKP2 NIPBL C5orf42 NUP155 WDR70 LIFR OSMR RICTOR TTC33 PRKAA1 RPL37 PLCXD3 OXCT1 C5orf51 SEPP1 ZNF131 HMGCS1 PAIP1 NNT HCN1 PARP8 ITGA2 MOCS2 NDUFS4 DHX29 SKIV2L2 PPAP2A IL6ST MAP3K1 GPBP1 PDE4D ELOVL7 SMIM15 ZSWIM6 DIMT1 KIF2A IPO11 RNF180 SREK1IP1 CWC27 PPWD1 TRIM23 NLN ERBB2IP SREK1 MAST4 PIK3R1 SLC30A5 RAD17 OCLN NAIP MAP1B MRPS27 TNPO1 BTF3 ANKRA2 GFM2 NSA2 FAM169A HMGCR COL4A3BP POLK PDE8B WDR41 TBCA AP3B1 SCAMP1 ARSB JMY HOMER1 PAPD4 CMYA5 THBS4 ZFYVE16 MTRNR2L2 MSH3 RASGRF2 SSBP2 RPS23 TMEM167A VCAN COX7C RASA1 CCNH TMEM161B LINC00461 MEF2C CETN3 LYSMD3 ARRDC3 FAM172A TTC37 RHOBTB3 GLRX ELL2 PCSK1 CAST LNPEP LIX1 RIOK2 CHD1 PAM PPIP5K2 EFNA5 FER PJA2 MAN2A1 WDR36 STARD4 NREP APC REEP5 DCP2 MCC YTHDC2 TRIM36 FEM1C CDO1 ATG12 AP3S1 COMMD10 SEMA6A DMXL1 HSD17B4 SNCAIP SNX2 SNX24 CSNK1G3 ZNF608 GRAMD3 ALDH7A1 PHAX PRRC1 SLC12A2 HINT1 LYRM7 CDC42SE2 FNIP1 ACSL6 P4HA2 IRF1 RAD50 KIF3A UQCRQ AFF4 ZCCHC10 HSPA4 C5orf15 VDAC1 SKP1 UBE2B CDKN2AIPNL SAR1B SEC24A CAMLG DDX46 TXNDC15 H2AFY CXCL14 SMAD5 SPOCK1 KLHL3 FAM13B BRD8 CDC23 KDM3B EGR1 ETF1 HSPA9 CTNNA1 SIL1 MATR3 PAIP2 DNAJC18 PURA CYSTM1 PFDN1 SRA1 SLC35A4 IK HARS HARS2 ZMAT2 TAF7 PCDHGC3 DIAPH1 HDAC3 KIAA0141 RNF14 GNPDA1 NDFIP1 SPRY4 FGF1 ARHGAP26 NR3C1 YIPF5 LARS RBM27 TCERG1 PPP2R2B STK32A DPYSL3 JAKMIP2 FBXO38 SH3TC2 GRPEL2 CSNK1A1 SLC26A2 CAMK2A TCOF1 RPS14 NDST1 RBM22 DCTN4 GPX3 CCDC69 GM2A SLC36A1 ATOX1 G3BP1 GRIA1 FAM114A2 MFAP3 LARP1 CNOT8 GEMIN5 MRPL22 SGCD HAVCR2 CYFIP2 ADAM19 CLINT1 EBF1 RNF145 UBLCP1 TTC1 PWWP2A SLU7 ATP10B GABRB2 GABRG2 CCNG1 NUDCD2 MAT2B TENM2 WWC1 RARS SLIT3 KCNIP1 NPM1 STK10 DUSP1 ERGIC1 RPL26L1 ATP6V0E1 CREBRF CPEB4 SFXN1 KIAA1191 HIGD2A FAF2 UIMC1 ZNF346 NSD1 LMAN2 DBN1 DDX41 TMED9 FAM153C NHP2 CLK4 RUFY1 HNRNPH1 CANX SQSTM1 GFPT2 CNOT6 LINC00847 BTNL9 TRIM7 GNB2L1 EXOC2 SERPINB1 SERPINB9 SERPINB6 NQO2 RIPK1 BPHL TUBB2A TUBB2B PRPF4B ECI2 CDYL LYRM4 NRN1 RIOK1 SNRNP48 PAK1IP1 TMEM14C TMEM14B ELOVL2 NEDD9 TMEM170B SIRT5 RANBP9 CD83 ATXN1 CAP2 NUP153aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0.81925 0.97575 0.60325 0.896 0.31125 0.736 0.9815 0.9905 0.95 0.6725 0.9975 0.99525 0.87575 0.97625 0.634 0.87275 0.7695 0.6315 0.65275 0.98775 0.97925 0.82025 0.9325 0.6655 0.95925 0.576 0.961 0.8425 0.73925 0.568 0.728 0.20475 0.66475 0.55925 0.89025 0.84775 0.62675 0.7335 0.70775 0.371 0.7565 0.983 0.74625 0.96275 0.76375 0.9815 0.767 0.77875 0.761 0.9665 0.353 0.76125 0.9605 0.99625 0.98475 0.9645 0.70725 0.88275 0.98725 0.99825 0.4775 0.88675 0.88125 0.96525DELETION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0GENES AFFECTED BY CNV: ZNF211 ZNF671 ZNF814 ZNF418 ZNF606 ZNF135 ZNF329 ZNF544 ZNF497 RPS5 CHMP2A SOX12 NRSN2 TBC1D20 CSNK2A1 PSMF1 FKBP1A NSFL1C STK35 SNRPB ZNF343 NOP56 IDH3B VPS16 PTPRA UBOX5 ITPA C20orf194 ATRN ADAM33 AP5S1 MAVS RNF24 SMOX PRNP RASSF2 SLC23A2 TMEM230 PCNA CDS2 GPCPD1 CRLS1 PLCB4 MKKS JAG1 BTBD3 ESF1 FLRT3 SNRPB2 PCSK2 DSTN SNX5 ZNF133 DZANK1 POLR3F RBBP9 SEC23B LINC00493 DTD1 RIN2 NAA20 CRNKL1 RALGAPA2 XRN2 CST3 APMAP ENTPD6 PYGB ZNF337 FRG1B HM13 BCL2L1 TM9SF4 POFUT1 KIF3B ASXL1 MAPRE1 CDK5RAP1 CBFA2T2 EIF2S2 AHCY ITCH DYNLRB1 PIGU NCOA6 GSS MYH7B EDEM2 EIF6 CEP250 ERGIC3 CPNE1 NFS1 ROMO1 RBM39 PHF20 LINC00657 EPB41L1 AAR2 DLGAP4 NDRG3 SOGA1 SAMHD1 RPN2 MANBAL BLCAP CTNNBL1 TTI1 KIAA1755 RALGAPB TOP1 ZHX3 CHD6 PTPRT L3MBTL1 IFT52 GDAP1L1 TTPAL SERINC3 PKIG YWHAB TOMM34 SDC4 SYS1 PIGT ACOT8 ZSWIM1 CTSA PLTP PCIF1 SLC35C2 ELMO2 ZNF334 TP53RK ZMYND8 SULF2 ARFGEF2 CSE1L STAU1 DDX27 ZNFX1 ZFAS1 B4GALT5 SLC9A8 RNF114 TMEM189 BCAS4 DPM1 ATP9A ZFP64 BCAS1 PFDN4 DOK5 FAM210B CSTF1 RTFDC1 BMP7 RAE1 RAB22A VAPB STX16 GNAS NELFCD ATP5E SLMO2 FAM217B PSMA7 SS18L1 OSBPL2 ADRM1 RPS21 GID8 NKAIN4 HELZ2 STMN3 TPD52L2 PRPF6 PCMTD2 CBS SMIM11 HSPA13 NRIP1aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0.96925 0.7375 0.94275 0.95175 0.9955 0.9915 0.65675 0.45175 0.9895 0.97775 0.87825 0.68175 0.99075 0.656 0.91625 0.952 0.95075 0.98175 0.955 0.36025 0.46875 0.8755 0.8195 0.77225 0.3435 0.97675 0.778 0.97425 0.98525 0.99225 0.90175 0.981 0.982 0.99625 0.9885 0.89825 0.97025 0.98175 0.9905 0.862 0.968 0.819 0.964 0.9435 0.93 0.74575 0.96925 0.809 0.976 0.79 0.99275 0.996 0.59225 0.56125 0.81725 0.8155 0.9825 0.8415 0.61475 0.71975 0.9865 0.95475 0.999 0.74275DELETION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0max vote:2GENES AFFECTED BY CNV: CRYL1 IFT88 XPO4 SAP18 SACS SPATA13 PARP4 MTMR6 NUPL1 RNF6 WASF3 GPR12 MTIF3 PAN3 FLT1 POMP KATNAL1 USPL1 HSPH1 FRY N4BP2L2 NBEA DCLK1 SPG20 SMAD9 ALG5 EXOSC8 SUPT20H UFM1 NHLRC3 LHFP COG6 FOXO1 MRPS31 KBTBD6 NAA16 AKAP11 DNAJC15 TSC22D1 GTF2F2 TPT1 ZC3H13 ESD SUCLA2 NUDT15 MED4 ITM2B LPAR6 RCBTB2 FNDC3A SETDB2 PHF11 RCBTB1 KPNA3 SPRYD7 FAM124A CKAP2 HNRNPA1L2 SUGT1 PCDH17 PCDH9 DIS3 PIBF1 KLF12 COMMD6 MYCBP2 EDNRB RBM26 NDFIP2 SPRY2 SLITRK1 GPR180 CLDN10 DZIP1 DNAJC3 UGGT2 MBNL2 IPO5 UBAC2 TM9SF2 TPP2 ARGLU1 CARKD ANKRD10 ARHGEF7 ATP11A PCID2 DCUN1D2 TMCO3 UPF3A TTC5 CCNB1IP1 PARP2 TEP1 OSGEP APEX1 METTL17 NDRG2 LINC00641 HNRNPC SUPT16H CHD8 RAB2B TOX4 METTL3 DAD1 ABHD4 OXA1L MRPL52 RBM23 PRMT5 PSMB5 ACIN1 C14orf119 SLC7A8 NGDN AP1G2 DHRS4L2 DCAF11 PSME1 PSME2 CHMP4A GMPR2 TINF2 RABGGTA DHRS1 SDR39U1 NOVA1 FOXG1 PRKD1 G2E3 SCFD1 STRN3 AP4S1 HECTD1 HEATR5A DTD2 NUBPL ARHGAP5 AKAP6 NPAS3 EGLN3 SPTSSA EAPP SNX6 CFL2 BAZ1A SRP54 FAM177A1 PPP2R3C NFKBIA RALGAPA1 BRMS1L MBIP SEC23A TRAPPC6B PNN KLHL28 FAM179B PRPF39 FKBP3 RPS29 RPL36AL NEMF ARF6 MAP4K5 ATL1 NIN PYGL TRIM9 TMX1 C14orf166 TXNDC16 ERO1L PSMC6 STYX GNPNAT1 FERMT2 DDHD1 GMFB CGRRF1 SAMD4A SOCS4 MAPK1IP1L LGALS3 ATG14 KTN1 EXOC5 AP5M1 ACTR10 PSMA3 ARID4A KIAA0586 DAAM1 JKAMP PCNXL4 DHRS7 PPM1A MNAT1 TRMT5 SLC38A6 HIF1A RHOJ SYNE2 MTHFD1 HSPA2 SPTB MAX FUT8 GPHN MPP5 ATP6V1D EIF2S1 PLEKHH1 VTI1B RDH11 ZFP36L1 ACTN1 DCAF5 EXD2 ERH SLC39A9 SRSF5 COX16 MED6 MAP3K9 PCNX SIPA1L1 RGS6 DPF3 ZFYVE1 RBM25 PSEN1 PAPLN NUMB ELMSAN1 FAM161B ALDH6A1 LIN52 ABCD4 NPC2 ISCA2 FCF1 RPS6KL1 EIF2B2 MLH3 NEK9 TMED10 C14orf1 TTLL5 TGFB3 IFT43 GPATCH2L ANGEL1 VIPAS39 AHSA1 SPTLC2 SLIRP SNW1 NRXN3 DIO2 STON2 SEL1L GALC KCNK10 ZC3H14 TTC8 FOXN3 TDP1 PSMC1 CALM1 C14orf159 SMEK1 FBLN5 TRIP11 ATXN3 NDUFB1 CPSF2 LGMN GOLGA5 UNC79 SERPINA3 CLMN C14orf132 GSKIP PAPOLA SETD3 YY1 WARS WDR25 MEG3 PPP2R5C DYNC1H1 HSP90AA1 WDR20 MOK CINP TECPR2 TNFAIP2 EIF5 MARK3 CKB BAG5 APOPT1 KLC1 ZFYVE21 C14orf2 INF2 SIVA1 PACS2 CRIP2aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0DELETION PROBABILITY: 0.92375 0.62125 0.99875 1 0.99925 0.9945 0.82 0.732 0.97525 0.9965 0.86475 0.917 1 0.91125 0.98325 0.4255 0.83725 0.998 0.9965 0.821 0.9155 0.9925 0.71 0.90325 0.826 0.8105 0.734 0.995 0.6085 0.9895 0.995 0.98875 0.9945 0.9975 0.88175 0.93525 0.7675 0.99775 0.9985 0.9935 0.995 0.453 0.99975 0.50175 0.99925 0.71775 0.993 0.53325 0.987 0.64375 1 0.99925 0.66575 0.91325 0.67475 0.88375 0.99175 0.99725 0.87925 0.452 0.99625 0.99375 0.9985 0.8025[1] "CNV SNPS:"
##   [1] "CRYL1"     "IFT88"     "XPO4"      "SAP18"     "SACS"     
##   [6] "SPATA13"   "PARP4"     "MTMR6"     "NUPL1"     "RNF6"     
##  [11] "WASF3"     "GPR12"     "MTIF3"     "PAN3"      "FLT1"     
##  [16] "POMP"      "KATNAL1"   "USPL1"     "HSPH1"     "FRY"      
##  [21] "N4BP2L2"   "NBEA"      "DCLK1"     "SPG20"     "SMAD9"    
##  [26] "ALG5"      "EXOSC8"    "SUPT20H"   "UFM1"      "NHLRC3"   
##  [31] "LHFP"      "COG6"      "FOXO1"     "MRPS31"    "KBTBD6"   
##  [36] "NAA16"     "AKAP11"    "DNAJC15"   "TSC22D1"   "GTF2F2"   
##  [41] "TPT1"      "ZC3H13"    "ESD"       "SUCLA2"    "NUDT15"   
##  [46] "MED4"      "ITM2B"     "LPAR6"     "RCBTB2"    "FNDC3A"   
##  [51] "SETDB2"    "PHF11"     "RCBTB1"    "KPNA3"     "SPRYD7"   
##  [56] "FAM124A"   "CKAP2"     "HNRNPA1L2" "SUGT1"     "PCDH17"   
##  [61] "PCDH9"     "DIS3"      "PIBF1"     "KLF12"     "COMMD6"   
##  [66] "MYCBP2"    "EDNRB"     "RBM26"     "NDFIP2"    "SPRY2"    
##  [71] "SLITRK1"   "GPR180"    "CLDN10"    "DZIP1"     "DNAJC3"   
##  [76] "UGGT2"     "MBNL2"     "IPO5"      "UBAC2"     "TM9SF2"   
##  [81] "TPP2"      "ARGLU1"    "CARKD"     "ANKRD10"   "ARHGEF7"  
##  [86] "ATP11A"    "PCID2"     "DCUN1D2"   "TMCO3"     "UPF3A"    
##  [91] "TTC5"      "CCNB1IP1"  "PARP2"     "TEP1"      "OSGEP"    
##  [96] "APEX1"     "METTL17"   "NDRG2"     "LINC00641" "HNRNPC"   
## [101] "SUPT16H"   "CHD8"      "RAB2B"     "TOX4"      "METTL3"   
## [106] "DAD1"      "ABHD4"     "OXA1L"     "MRPL52"    "RBM23"    
## [111] "PRMT5"     "PSMB5"     "ACIN1"     "C14orf119" "SLC7A8"   
## [116] "NGDN"      "AP1G2"     "DHRS4L2"   "DCAF11"    "PSME1"    
## [121] "PSME2"     "CHMP4A"    "GMPR2"     "TINF2"     "RABGGTA"  
## [126] "DHRS1"     "SDR39U1"   "NOVA1"     "FOXG1"     "PRKD1"    
## [131] "G2E3"      "SCFD1"     "STRN3"     "AP4S1"     "HECTD1"   
## [136] "HEATR5A"   "DTD2"      "NUBPL"     "ARHGAP5"   "AKAP6"    
## [141] "NPAS3"     "EGLN3"     "SPTSSA"    "EAPP"      "SNX6"     
## [146] "CFL2"      "BAZ1A"     "SRP54"     "FAM177A1"  "PPP2R3C"  
## [151] "NFKBIA"    "RALGAPA1"  "BRMS1L"    "MBIP"      "SEC23A"   
## [156] "TRAPPC6B"  "PNN"       "KLHL28"    "FAM179B"   "PRPF39"   
## [161] "FKBP3"     "RPS29"     "RPL36AL"   "NEMF"      "ARF6"     
## [166] "MAP4K5"    "ATL1"      "NIN"       "PYGL"      "TRIM9"    
## [171] "TMX1"      "C14orf166" "TXNDC16"   "ERO1L"     "PSMC6"    
## [176] "STYX"      "GNPNAT1"   "FERMT2"    "DDHD1"     "GMFB"     
## [181] "CGRRF1"    "SAMD4A"    "SOCS4"     "MAPK1IP1L" "LGALS3"   
## [186] "ATG14"     "KTN1"      "EXOC5"     "AP5M1"     "ACTR10"   
## [191] "PSMA3"     "ARID4A"    "KIAA0586"  "DAAM1"     "JKAMP"    
## [196] "PCNXL4"    "DHRS7"     "PPM1A"     "MNAT1"     "TRMT5"    
## [201] "SLC38A6"   "HIF1A"     "RHOJ"      "SYNE2"     "MTHFD1"   
## [206] "HSPA2"     "SPTB"      "MAX"       "FUT8"      "GPHN"     
## [211] "MPP5"      "ATP6V1D"   "EIF2S1"    "PLEKHH1"   "VTI1B"    
## [216] "RDH11"     "ZFP36L1"   "ACTN1"     "DCAF5"     "EXD2"     
## [221] "ERH"       "SLC39A9"   "SRSF5"     "COX16"     "MED6"     
## [226] "MAP3K9"    "PCNX"      "SIPA1L1"   "RGS6"      "DPF3"     
## [231] "ZFYVE1"    "RBM25"     "PSEN1"     "PAPLN"     "NUMB"     
## [236] "ELMSAN1"   "FAM161B"   "ALDH6A1"   "LIN52"     "ABCD4"    
## [241] "NPC2"      "ISCA2"     "FCF1"      "RPS6KL1"   "EIF2B2"   
## [246] "MLH3"      "NEK9"      "TMED10"    "C14orf1"   "TTLL5"    
## [251] "TGFB3"     "IFT43"     "GPATCH2L"  "ANGEL1"    "VIPAS39"  
## [256] "AHSA1"     "SPTLC2"    "SLIRP"     "SNW1"      "NRXN3"    
## [261] "DIO2"      "STON2"     "SEL1L"     "GALC"      "KCNK10"   
## [266] "ZC3H14"    "TTC8"      "FOXN3"     "TDP1"      "PSMC1"    
## [271] "CALM1"     "C14orf159" "SMEK1"     "FBLN5"     "TRIP11"   
## [276] "ATXN3"     "NDUFB1"    "CPSF2"     "LGMN"      "GOLGA5"   
## [281] "UNC79"     "SERPINA3"  "CLMN"      "C14orf132" "GSKIP"    
## [286] "PAPOLA"    "SETD3"     "YY1"       "WARS"      "WDR25"    
## [291] "MEG3"      "PPP2R5C"   "DYNC1H1"   "HSP90AA1"  "WDR20"    
## [296] "MOK"       "CINP"      "TECPR2"    "TNFAIP2"   "EIF5"     
## [301] "MARK3"     "CKB"       "BAG5"      "APOPT1"    "KLC1"     
## [306] "ZFYVE21"   "C14orf2"   "INF2"      "SIVA1"     "PACS2"    
## [311] "CRIP2"    
##      MGH31_A02 MGH31_A03 MGH31_A05 MGH31_A06 MGH31_A07 MGH31_A10 MGH31_A11
## [1,]   0.92375   0.62125   0.99875         1   0.99925    0.9945      0.82
##      MGH31_B01 MGH31_B02 MGH31_B04 MGH31_B06 MGH31_B08 MGH31_B09 MGH31_B10
## [1,]     0.732   0.97525    0.9965   0.86475     0.917         1   0.91125
##      MGH31_C02 MGH31_C03 MGH31_C04 MGH31_C05 MGH31_C06 MGH31_C07 MGH31_C09
## [1,]   0.98325    0.4255   0.83725     0.998    0.9965     0.821    0.9155
##      MGH31_C10 MGH31_C11 MGH31_D03 MGH31_D05 MGH31_D07 MGH31_D09 MGH31_D10
## [1,]    0.9925      0.71   0.90325     0.826    0.8105     0.734     0.995
##      MGH31_D11 MGH31_D12 MGH31_E01 MGH31_E02 MGH31_E03 MGH31_E05 MGH31_E08
## [1,]    0.6085    0.9895     0.995   0.98875    0.9945    0.9975   0.88175
##      MGH31_E11 MGH31_E12 MGH31_F02 MGH31_F03 MGH31_F04 MGH31_F06 MGH31_F07
## [1,]   0.93525    0.7675   0.99775    0.9985    0.9935     0.995     0.453
##      MGH31_F08 MGH31_F09 MGH31_F11 MGH31_F12 MGH31_G01 MGH31_G02 MGH31_G05
## [1,]   0.99975   0.50175   0.99925   0.71775     0.993   0.53325     0.987
##      MGH31_G07 MGH31_G08 MGH31_G09 MGH31_G10 MGH31_G12 MGH31_H01 MGH31_H02
## [1,]   0.64375         1   0.99925   0.66575   0.91325   0.67475   0.88375
##      MGH31_H04 MGH31_H05 MGH31_H06 MGH31_H08 MGH31_H09 MGH31_H10 MGH31_H11
## [1,]   0.99175   0.99725   0.87925     0.452   0.99625   0.99375    0.9985
##      MGH31_H12
## [1,]    0.8025
## [1] "GROUP1:"
##  [1] "MGH31_A02" "MGH31_A05" "MGH31_A06" "MGH31_A07" "MGH31_A10"
##  [6] "MGH31_A11" "MGH31_B02" "MGH31_B04" "MGH31_B06" "MGH31_B08"
## [11] "MGH31_B09" "MGH31_B10" "MGH31_C02" "MGH31_C04" "MGH31_C05"
## [16] "MGH31_C06" "MGH31_C07" "MGH31_C09" "MGH31_C10" "MGH31_D03"
## [21] "MGH31_D05" "MGH31_D07" "MGH31_D10" "MGH31_D12" "MGH31_E01"
## [26] "MGH31_E02" "MGH31_E03" "MGH31_E05" "MGH31_E08" "MGH31_E11"
## [31] "MGH31_E12" "MGH31_F02" "MGH31_F03" "MGH31_F04" "MGH31_F06"
## [36] "MGH31_F08" "MGH31_F11" "MGH31_G01" "MGH31_G05" "MGH31_G08"
## [41] "MGH31_G09" "MGH31_G12" "MGH31_H02" "MGH31_H04" "MGH31_H05"
## [46] "MGH31_H06" "MGH31_H09" "MGH31_H10" "MGH31_H11" "MGH31_H12"
## [1] "GROUP2:"
## character(0)
## [1] "Recursion for Group1"
## max vote:2GENES AFFECTED BY CNV: NSUN2 ADCY2 MTRR SEMA5A FAM173B CCT5 CMBL MARCH6 CTNND2 TRIO ANKH FAM134B MYO10 BASP1 GUSBP1 CDH10 CDH6 DROSHA C5orf22 PDZD2 MTMR12 ZFR SUB1 TARS C1QTNF3 RAI14 RAD1 BRIX1 LMBRD2 SKP2 NIPBL C5orf42 NUP155 WDR70 LIFR OSMR RICTOR TTC33 PRKAA1 RPL37 PLCXD3 OXCT1 C5orf51 SEPP1 ZNF131 HMGCS1 PAIP1 NNT HCN1 PARP8 ITGA2 MOCS2 NDUFS4 DHX29 SKIV2L2 PPAP2A IL6ST MAP3K1 GPBP1 PDE4D ELOVL7 SMIM15 ZSWIM6 DIMT1 KIF2A IPO11 RNF180 SREK1IP1 CWC27 PPWD1 TRIM23 NLN ERBB2IP SREK1 MAST4 PIK3R1 SLC30A5 RAD17 OCLN NAIP MAP1B MRPS27 TNPO1 BTF3 ANKRA2 GFM2 NSA2 FAM169A HMGCR COL4A3BP POLK PDE8B WDR41 TBCA AP3B1 SCAMP1 ARSB JMY HOMER1 PAPD4 CMYA5 THBS4 ZFYVE16 MTRNR2L2 MSH3 RASGRF2 SSBP2 RPS23 TMEM167A VCAN COX7C RASA1 CCNH TMEM161B LINC00461 MEF2C CETN3 LYSMD3 ARRDC3 FAM172A TTC37 RHOBTB3 GLRX ELL2 PCSK1 CAST LNPEP LIX1 RIOK2 CHD1 PAM PPIP5K2 EFNA5 FER PJA2 MAN2A1 WDR36 STARD4 NREP APC REEP5 DCP2 MCC YTHDC2 TRIM36 FEM1C CDO1 ATG12 AP3S1 COMMD10 SEMA6A DMXL1 HSD17B4 SNCAIP SNX2 SNX24 CSNK1G3 ZNF608 GRAMD3 ALDH7A1 PHAX PRRC1 SLC12A2 HINT1 LYRM7 CDC42SE2 FNIP1 ACSL6 P4HA2 IRF1 RAD50 KIF3A UQCRQ AFF4 ZCCHC10 HSPA4 C5orf15 VDAC1 SKP1 UBE2B CDKN2AIPNL SAR1B SEC24A CAMLG DDX46 TXNDC15 H2AFY CXCL14 SMAD5 SPOCK1 KLHL3 FAM13B BRD8 CDC23 KDM3B EGR1 ETF1 HSPA9 CTNNA1 SIL1 MATR3 PAIP2 DNAJC18 PURA CYSTM1 PFDN1 SRA1 SLC35A4 IK HARS HARS2 ZMAT2 TAF7 PCDHGC3 DIAPH1 HDAC3 KIAA0141 RNF14 GNPDA1 NDFIP1 SPRY4 FGF1 ARHGAP26 NR3C1 YIPF5 LARS RBM27 TCERG1 PPP2R2B STK32A DPYSL3 JAKMIP2 FBXO38 SH3TC2 GRPEL2 CSNK1A1 SLC26A2 CAMK2A TCOF1 RPS14 NDST1 RBM22 DCTN4 GPX3 CCDC69 GM2A SLC36A1 ATOX1 G3BP1 GRIA1 FAM114A2 MFAP3 LARP1 CNOT8 GEMIN5 MRPL22 SGCD HAVCR2 CYFIP2 ADAM19 CLINT1 EBF1 RNF145 UBLCP1 TTC1 PWWP2A SLU7 ATP10B GABRB2 GABRG2 CCNG1 NUDCD2 MAT2B TENM2 WWC1 RARS SLIT3 KCNIP1 NPM1 STK10 DUSP1 ERGIC1 RPL26L1 ATP6V0E1 CREBRF CPEB4 SFXN1 KIAA1191 HIGD2A FAF2 UIMC1 ZNF346 NSD1 LMAN2 DBN1 DDX41 TMED9 FAM153C NHP2 CLK4 RUFY1 HNRNPH1 CANX SQSTM1 GFPT2 CNOT6 LINC00847 BTNL9 TRIM7 GNB2L1 EXOC2 SERPINB1 SERPINB9 SERPINB6 NQO2 RIPK1 BPHL TUBB2A TUBB2B PRPF4B ECI2 CDYL LYRM4 NRN1 RIOK1 SNRNP48 PAK1IP1 TMEM14C TMEM14B ELOVL2 NEDD9 TMEM170B SIRT5 RANBP9 CD83 ATXN1 CAP2 NUP153aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0.823 0.6455 0.88825 0.2825 0.738 0.984 0.9575 0.69825 0.99875 0.995 0.87275 0.98025 0.6775 0.7675 0.65925 0.68525 0.9835 0.983 0.84775 0.66825 0.973 0.61 0.83275 0.6025 0.73275 0.22425 0.70825 0.5895 0.881 0.836 0.631 0.72925 0.68925 0.3605 0.76775 0.759 0.75925 0.7715 0.762 0.355 0.735 0.99725 0.956 0.769 0.888 0.979 0.49425 0.887 0.87475 0.96075DELETION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0GENES AFFECTED BY CNV: ZNF211 ZNF671 ZNF814 ZNF418 ZNF606 ZNF135 ZNF329 ZNF544 ZNF497 RPS5 CHMP2A SOX12 NRSN2 TBC1D20 CSNK2A1 PSMF1 FKBP1A NSFL1C STK35 SNRPB ZNF343 NOP56 IDH3B VPS16 PTPRA UBOX5 ITPA C20orf194 ATRN ADAM33 AP5S1 MAVS RNF24 SMOX PRNP RASSF2 SLC23A2 TMEM230 PCNA CDS2 GPCPD1 CRLS1 PLCB4 MKKS JAG1 BTBD3 ESF1 FLRT3 SNRPB2 PCSK2 DSTN SNX5 ZNF133 DZANK1 POLR3F RBBP9 SEC23B LINC00493 DTD1 RIN2 NAA20 CRNKL1 RALGAPA2 XRN2 CST3 APMAP ENTPD6 PYGB ZNF337 FRG1B HM13 BCL2L1 TM9SF4 POFUT1 KIF3B ASXL1 MAPRE1 CDK5RAP1 CBFA2T2 EIF2S2 AHCY ITCH DYNLRB1 PIGU NCOA6 GSS MYH7B EDEM2 EIF6 CEP250 ERGIC3 CPNE1 NFS1 ROMO1 RBM39 PHF20 LINC00657 EPB41L1 AAR2 DLGAP4 NDRG3 SOGA1 SAMHD1 RPN2 MANBAL BLCAP CTNNBL1 TTI1 KIAA1755 RALGAPB TOP1 ZHX3 CHD6 PTPRT L3MBTL1 IFT52 GDAP1L1 TTPAL SERINC3 PKIG YWHAB TOMM34 SDC4 SYS1 PIGT ACOT8 ZSWIM1 CTSA PLTP PCIF1 SLC35C2 ELMO2 ZNF334 TP53RK ZMYND8 SULF2 ARFGEF2 CSE1L STAU1 DDX27 ZNFX1 ZFAS1 B4GALT5 SLC9A8 RNF114 TMEM189 BCAS4 DPM1 ATP9A ZFP64 BCAS1 PFDN4 DOK5 FAM210B CSTF1 RTFDC1 BMP7 RAE1 RAB22A VAPB STX16 GNAS NELFCD ATP5E SLMO2 FAM217B PSMA7 SS18L1 OSBPL2 ADRM1 RPS21 GID8 NKAIN4 HELZ2 STMN3 TPD52L2 PRPF6 PCMTD2 CBS SMIM11 HSPA13 NRIP1aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0.97475 0.95475 0.94925 0.99825 0.9895 0.64725 0.99075 0.97725 0.882 0.6845 0.989 0.62 0.91375 0.94975 0.982 0.9465 0.37225 0.48425 0.87275 0.777 0.32925 0.97275 0.981 0.98825 0.8965 0.975 0.9875 0.99625 0.99125 0.91725 0.97175 0.986 0.99275 0.853 0.96425 0.96675 0.935 0.97425 0.98075 0.98825 0.99575 0.55825 0.81525 0.98075 0.8445 0.62475 0.98625 0.95375 0.998 0.752DELETION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0max vote:2GENES AFFECTED BY CNV: CDC5L RCAN2 MUT MCM3 PAQR8 EFHC1 TRAM2 TMEM14A GSTA4 ICK FBXO9 ELOVL5 LRRC1 DST ZNF451 PTP4A1 PHF3 LMBRD1 FAM135A RIMS1 MTO1 SLC17A5 COX7A2 TMEM30A SENP6 MYO6 PHIP HMGN3 SH3BGRL2 BCKDHB IBTK PGM3 MRAP2 SNX14 SYNCRIP SNHG5 ZNF292 RARS2 ORC3 CNR1 RNGTT PNRC1 SRSF12 PM20D2 RRAGD MDN1 CASP8AP2 BACH2 MAP3K7 MANEA FUT9 UFL1 NDUFAF4 POU3F2 PNISR USP45 CCNC ASCC3 GRIK2 BVES ATG5 QRSL1 C6orf203 PDSS2 SOBP SEC63 OSTM1 SNX3 FOXO3 CD164 FIG4 CDC40 CDK19 AMD1 GTF3C6 RPF2 REV3L TRAF3IP2 FYN TUBE1 FAM229B LAMA4 MARCKS HDAC2 NT5DC1 TSPYL4 TSPYL1 RWDD1 SLC35F1 CEP85L ASF1A GJA1 HSF2 SERINC1 PKIB FABP7 HDDC2 NCOA7 HINT3 TRMT11 ECHDC1 LAMA2 EPB41L2 AKAP7 MED23 STX7 SLC18B1 RPS12 SGK1 HBS1L AHI1 PDE7B BCLAF1 IFNGR1 HECA VTA1 AIG1 ADAT2 PEX3 FUCA2 LTV1 SF3B5 UTRN SHPRH SASH1 UST TAB2 PPIL4 LATS1 NUP43 PCMT1 PLEKHG1 MTHFD1L AKAP12 RMND1 SYNE1 MTRF1L RGS17 IPCEF1 ARID1B TMEM242 SERAC1 GTF2H5 TULP4 TMEM181 DYNLT1 EZR WTAP ACAT2 TCP1 MRPL18 IGF2R PARK2 QKI SFT2D1 MPC1 RPS6KA2 MLLT4 THBS2 WDR27 TBP CSMD1 MCPH1 AGPAT5 MFHAS1 ERI1 TNKS MSRA MTMR9 NEIL2 FDFT1 CTSB KIAA1456 DLC1 ZDHHC2 CNOT7 SLC7A2 MTUS1 PCM1 ASAH1 PSD3 CSGALNACT1 INTS10 LPL ATP6V1B2 LZTS1 SLC39A14 EGR3 CHMP7 R3HCC1 ENTPD4 DOCK5 PPP2R2A BNIP3L DPYSL2 TRIM35 PTK2B SCARA3 CCDC25 ELP3 FZD3 EXTL3 INTS9 HMBOX1 KIF13B LEPROTL1 DCTN6 RBPMS UBXN8 WRN ERLIN2 PROSC ASH2L LSM1 BAG4 DDHD2 PPAPDC1B WHSC1L1 FGFR1 TACC1 PLEKHA2 TM2D2 GOLGA7 AGPAT6 ANK1 KAT6A AP3M2 PLAT VDAC3 SLC20A2 RNF170 FNTA HGSNAT PRKDC MCM4 UBE2V2 ST18 RB1CC1 ATP6V1H TCEA1 LYPLA1 MRPL15 TMEM68 TGS1 RPS20 CHCHD7 IMPAD1 UBXN2B SDCBP NSMAF TOX RAB2A CHD7 ASPH GGH YTHDF3 ARMC1 C8orf46 VCPIP1 SNHG6 COPS5 CSPP1 ARFGEF1 PREX2 SULF1 TRAM1 TERF1 RPL7 RDH10 STAU2 UBE2W TCEB1 TMEM70 GDAP1 ZFHX4 PEX2 STMN2 HEY1 MRPS28 TPD52 ZNF704 PAG1 PMP2 IMPA1 ZFAND1 RALYL LRRCC1 C8orf59 CA2 WWP1 RMDN1 CPNE3 MMP16 NBN DECR1 TMEM55A OTUD6B RUNX1T1 FAM92A1 TMEM67 PDP1 KIAA1429 DPY19L4 INTS8 UQCRB PTDSS1 CPQ MTDH LAPTM4B MATN2 RPL30 HRSP12 VPS13B COX6C POLR2K SPAG1 RNF19A ANKRD46 PABPC1 YWHAZ ZNF706 NCALD RRM2B UBR5 KLF10 AZIN1 ATP6V1C1 BAALC SLC25A32 DCAF13 RIMS2 OXR1 EIF3E EMC2 NUDCD1 ENY2 SYBU CSMD3 TRPS1 EIF3H UTP23 RAD21 SAMD12 ENPP2 TAF2 MRPL13 DERL1 ZHX1 ATAD2 FBXO32 NDUFB9 SQLE KIAA0196 FAM84B FAM49B ASAP1 PHF20L1 NDRG1 ST3GAL1 KHDRBS3 CHRAC1 PTK2 EEF1D KIAA1875 VPS28 RPL8 ZNF517 ZNF250 ZNF252P C8orf33 KANK1 SMARCA2 RFX3 GLIS3 CDC37L1 AK3 KIAA2026 RANBP6 IL33 GLDC KDM4C PTPRD MPDZ ZDHHC21 SNAPC3 PSIP1 RPS6 SLC24A2 MLLT3 FOCAD KLHL9 MTAP IFT74 MOB3B C9orf72 ACO1 DDX58 NDUFB6 APTX DNAJA1 SMU1 B4GALT1 BAG1 CHMP5 NFX1 NOL6 UBE2R2 UBAP2 DCAF12 UBAP1 KIAA1161 DCTN3 GALT VCP PIGO STOML2 UNC13B TLN1 CREB3 GBA2 GLIPR2 GNE RNF38 ZCCHC7 GRHPR POLR1E TRMT10B DCAF10aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0DELETION PROBABILITY: 0.794 0.56475 0.6135 0.91075 0.573 0.24125 0.79775 0.77225 0.5095 0.42175 0.7735 0.83875 0.79525 0.42225 0.2435 0.65125 0.522 0.64025 0.87325 0.2145 0.56425 0.68 0.99725 0.99625 0.76425 0.81775 0.69575 0.709 0.39475 0.43575 0.7495 0.999 0.77325 0.91525 0.6845 0.85175 0.67425 0.69125 0.88875 0.7905 0.6355 0.83425 0.418 0.97475 0.936 0.67425 0.37 0.98225 0.19 0.2765GENES AFFECTED BY CNV: CD82 SYT13 CHST1 PHF21A MDK AMBRA1 ATG13 ARHGAP1 CKAP5 LRP4 C11orf49 ARFGAP2 DDB2 ACP2 PSMC3 CELF1 NDUFS3 MTCH2 FNBP4 NUP160 PTPRJ SSRP1 TIMM10 UBE2L6 SERPING1 ZDHHC5 TMX2 C11orf31 FAM111A STX3 MRPL16 DDB1 CPSF7 TMEM258 FEN1 FADS1 FADS2 ASRGL1 AHNAK TUT1 B3GAT3 GANAB UBXN1 GNG3 TTC9C POLR2G NXF1 SLC3A2 RARRES3 PLA2G16 ATL3 RTN3 NAA40 COX8A OTUB1 STIP1 TRPT1 FKBP2 TRMT112 PRDX5 MEN1 ARL2 FAU DPF2 FRMD8 NEAT1 PCNXL3 KAT5 RNASEH2C AP5B1 CFL1 EFEMP2 FIBP DRAP1 SART1 EIF1AD BANF1 SF3B2 RAB1B YIF1A BRMS1 MRPL11 DPP3 CCS RBM14 RBM4B C11orf80 KDM2A POLD4 PPP1CA AIP CDK2AP2 GSTP1 NDUFV1 NDUFS8 SUV420H1 C11orf24 PPP6R3 CPT1A MRPL21 ORAOV1 PPFIA1 CTTN DHCR7 NADSYN1 RNF121 NUMA1 LAMTOR1 LRTOMT ARAP1 FCHSD2 PLEKHB1 RAB6A MRPL48 COA4 PAAF1 UCP2 PPME1 PGM2L1 SPCS2 ARRB1 RPS3 ACER3 PAK1 CLNS1A RSF1 INTS4 ALG8 KCTD21 TENM4 PRCP RAB30 PCF11 ANKRD42 CCDC90B DLG2 TMEM126B TMEM126A CREBZF SYTL2 PICALM C11orf73 TMEM135 CTSC NAALAD2 CHORDC1 FAT3 C11orf54 TAF1D MED17 GPR83 AMOTL1 CWC15 ENDOD1 SESN3 FAM76B CEP57 MTMR2 CCDC82 JRKL YAP1 BIRC2 TMEM123 DCUN1D5 DYNC2H1 GRIA4 MSANTD4 AASDHPPT CWF19L2 ELMOD1 CUL5 ACAT1 NPAT ATM KDELC2 DDX10 ZC3H12C RDX FDX1 LAYN PPP2R1B ALG9 CRYAB DIXDC1 DLAT SDHD BCO2 NCAM1 USP28 REXO2 CADM1 SIK3 TAGLN PCSK7 BACE1aggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0DELETION PROBABILITY: 0.4235 0.47525 0.142 0.76275 0.50575 0.95925 0.9695 0.65 0.99475 0.9395 0.70775 0.8815 0.5905 0.837 0.84375 0.74475 0.976 0.988 0.934 0.91775 0.97025 0.73875 0.766 0.28675 0.9915 0.55375 0.906 0.7915 0.86675 0.74525 0.87725 0.7605 0.5945 0.946 0.5425 0.535 0.6205 0.55825 0.6955 0.3755 0.64525 0.99475 0.9625 0.7545 0.49325 0.95825 0.90375 0.9565 0.91175 0.9505[1] "CNV SNPS:"
##   [1] "ZNF211"    "ZNF671"    "ZNF814"    "ZNF418"    "ZNF606"   
##   [6] "ZNF135"    "ZNF329"    "ZNF544"    "ZNF497"    "RPS5"     
##  [11] "CHMP2A"    "SOX12"     "NRSN2"     "TBC1D20"   "CSNK2A1"  
##  [16] "PSMF1"     "FKBP1A"    "NSFL1C"    "STK35"     "SNRPB"    
##  [21] "ZNF343"    "NOP56"     "IDH3B"     "VPS16"     "PTPRA"    
##  [26] "UBOX5"     "ITPA"      "C20orf194" "ATRN"      "ADAM33"   
##  [31] "AP5S1"     "MAVS"      "RNF24"     "SMOX"      "PRNP"     
##  [36] "RASSF2"    "SLC23A2"   "TMEM230"   "PCNA"      "CDS2"     
##  [41] "GPCPD1"    "CRLS1"     "PLCB4"     "MKKS"      "JAG1"     
##  [46] "BTBD3"     "ESF1"      "FLRT3"     "SNRPB2"    "PCSK2"    
##  [51] "DSTN"      "SNX5"      "ZNF133"    "DZANK1"    "POLR3F"   
##  [56] "RBBP9"     "SEC23B"    "LINC00493" "DTD1"      "RIN2"     
##  [61] "NAA20"     "CRNKL1"    "RALGAPA2"  "XRN2"      "CST3"     
##  [66] "APMAP"     "ENTPD6"    "PYGB"      "ZNF337"    "FRG1B"    
##  [71] "HM13"      "BCL2L1"    "TM9SF4"    "POFUT1"    "KIF3B"    
##  [76] "ASXL1"     "MAPRE1"    "CDK5RAP1"  "CBFA2T2"   "EIF2S2"   
##  [81] "AHCY"      "ITCH"      "DYNLRB1"   "PIGU"      "NCOA6"    
##  [86] "GSS"       "MYH7B"     "EDEM2"     "EIF6"      "CEP250"   
##  [91] "ERGIC3"    "CPNE1"     "NFS1"      "ROMO1"     "RBM39"    
##  [96] "PHF20"     "LINC00657" "EPB41L1"   "AAR2"      "DLGAP4"   
## [101] "NDRG3"     "SOGA1"     "SAMHD1"    "RPN2"      "MANBAL"   
## [106] "BLCAP"     "CTNNBL1"   "TTI1"      "KIAA1755"  "RALGAPB"  
## [111] "TOP1"      "ZHX3"      "CHD6"      "PTPRT"     "L3MBTL1"  
## [116] "IFT52"     "GDAP1L1"   "TTPAL"     "SERINC3"   "PKIG"     
## [121] "YWHAB"     "TOMM34"    "SDC4"      "SYS1"      "PIGT"     
## [126] "ACOT8"     "ZSWIM1"    "CTSA"      "PLTP"      "PCIF1"    
## [131] "SLC35C2"   "ELMO2"     "ZNF334"    "TP53RK"    "ZMYND8"   
## [136] "SULF2"     "ARFGEF2"   "CSE1L"     "STAU1"     "DDX27"    
## [141] "ZNFX1"     "ZFAS1"     "B4GALT5"   "SLC9A8"    "RNF114"   
## [146] "TMEM189"   "BCAS4"     "DPM1"      "ATP9A"     "ZFP64"    
## [151] "BCAS1"     "PFDN4"     "DOK5"      "FAM210B"   "CSTF1"    
## [156] "RTFDC1"    "BMP7"      "RAE1"      "RAB22A"    "VAPB"     
## [161] "STX16"     "GNAS"      "NELFCD"    "ATP5E"     "SLMO2"    
## [166] "FAM217B"   "PSMA7"     "SS18L1"    "OSBPL2"    "ADRM1"    
## [171] "RPS21"     "GID8"      "NKAIN4"    "HELZ2"     "STMN3"    
## [176] "TPD52L2"   "PRPF6"     "PCMTD2"    "CBS"       "SMIM11"   
## [181] "HSPA13"    "NRIP1"    
##      MGH31_A02 MGH31_A05 MGH31_A06 MGH31_A07 MGH31_A10 MGH31_A11 MGH31_B02
## [1,]   0.97475   0.95475   0.94925   0.99825    0.9895   0.64725   0.99075
##      MGH31_B04 MGH31_B06 MGH31_B08 MGH31_B09 MGH31_B10 MGH31_C02 MGH31_C04
## [1,]   0.97725     0.882    0.6845     0.989      0.62   0.91375   0.94975
##      MGH31_C05 MGH31_C06 MGH31_C07 MGH31_C09 MGH31_C10 MGH31_D03 MGH31_D05
## [1,]     0.982    0.9465   0.37225   0.48425   0.87275     0.777   0.32925
##      MGH31_D07 MGH31_D10 MGH31_D12 MGH31_E01 MGH31_E02 MGH31_E03 MGH31_E05
## [1,]   0.97275     0.981   0.98825    0.8965     0.975    0.9875   0.99625
##      MGH31_E08 MGH31_E11 MGH31_E12 MGH31_F02 MGH31_F03 MGH31_F04 MGH31_F06
## [1,]   0.99125   0.91725   0.97175     0.986   0.99275     0.853   0.96425
##      MGH31_F08 MGH31_F11 MGH31_G01 MGH31_G05 MGH31_G08 MGH31_G09 MGH31_G12
## [1,]   0.96675     0.935   0.97425   0.98075   0.98825   0.99575   0.55825
##      MGH31_H02 MGH31_H04 MGH31_H05 MGH31_H06 MGH31_H09 MGH31_H10 MGH31_H11
## [1,]   0.81525   0.98075    0.8445   0.62475   0.98625   0.95375     0.998
##      MGH31_H12
## [1,]     0.752
## [1] "GROUP1:"
##  [1] "MGH31_A02" "MGH31_A05" "MGH31_A06" "MGH31_A07" "MGH31_A10"
##  [6] "MGH31_B02" "MGH31_B04" "MGH31_B06" "MGH31_B09" "MGH31_C02"
## [11] "MGH31_C04" "MGH31_C05" "MGH31_C06" "MGH31_C10" "MGH31_D03"
## [16] "MGH31_D07" "MGH31_D10" "MGH31_D12" "MGH31_E01" "MGH31_E02"
## [21] "MGH31_E03" "MGH31_E05" "MGH31_E08" "MGH31_E11" "MGH31_E12"
## [26] "MGH31_F02" "MGH31_F03" "MGH31_F04" "MGH31_F06" "MGH31_F08"
## [31] "MGH31_F11" "MGH31_G01" "MGH31_G05" "MGH31_G08" "MGH31_G09"
## [36] "MGH31_H02" "MGH31_H04" "MGH31_H05" "MGH31_H09" "MGH31_H10"
## [41] "MGH31_H11" "MGH31_H12"
## [1] "GROUP2:"
## character(0)
## [1] "Recursion for Group1"
## max vote:0max vote:2GENES AFFECTED BY CNV: MUT MCM3 PAQR8 EFHC1 TRAM2 TMEM14A GSTA4 ICK FBXO9 ELOVL5 LRRC1 DST ZNF451 PTP4A1 PHF3 LMBRD1 FAM135A RIMS1 MTO1 SLC17A5 COX7A2 TMEM30A SENP6 MYO6 PHIP HMGN3 SH3BGRL2 BCKDHB IBTK PGM3 MRAP2 SNX14 SYNCRIP SNHG5 ZNF292 RARS2 ORC3 CNR1 RNGTT PNRC1 SRSF12 PM20D2 RRAGD MDN1 CASP8AP2 BACH2 MAP3K7 MANEA FUT9 UFL1 NDUFAF4 POU3F2 PNISR USP45 CCNC ASCC3 GRIK2 BVES ATG5 QRSL1 C6orf203 PDSS2 SOBP SEC63 OSTM1 SNX3 FOXO3 CD164 FIG4 CDC40 CDK19 AMD1 GTF3C6 RPF2 REV3L TRAF3IP2 FYN TUBE1 FAM229B LAMA4 MARCKS HDAC2 NT5DC1 TSPYL4 TSPYL1 RWDD1 SLC35F1 CEP85L ASF1A GJA1 HSF2 SERINC1 PKIB FABP7 HDDC2 NCOA7 HINT3 TRMT11 ECHDC1 LAMA2 EPB41L2 AKAP7 MED23 STX7 SLC18B1 RPS12 SGK1 HBS1L AHI1 PDE7B BCLAF1 IFNGR1 HECA VTA1 AIG1 ADAT2 PEX3 FUCA2 LTV1 SF3B5 UTRN SHPRH SASH1 UST TAB2 PPIL4 LATS1 NUP43 PCMT1 PLEKHG1 MTHFD1L AKAP12 RMND1 SYNE1 MTRF1L RGS17 IPCEF1 ARID1B TMEM242 SERAC1 GTF2H5 TULP4 TMEM181 DYNLT1 EZR WTAP ACAT2 TCP1 MRPL18 IGF2R PARK2 QKI SFT2D1 MPC1 RPS6KA2 MLLT4 THBS2 WDR27 TBP CSMD1 MCPH1 AGPAT5 MFHAS1 ERI1 TNKS MSRA MTMR9 NEIL2 FDFT1 CTSB KIAA1456 DLC1 ZDHHC2 CNOT7 SLC7A2 MTUS1 PCM1 ASAH1 PSD3 CSGALNACT1 INTS10 LPL ATP6V1B2 LZTS1 SLC39A14 EGR3 CHMP7 R3HCC1 ENTPD4 DOCK5 PPP2R2A BNIP3L DPYSL2 TRIM35 PTK2B SCARA3 CCDC25 ELP3 FZD3 EXTL3 INTS9 HMBOX1 KIF13B LEPROTL1 DCTN6 RBPMS UBXN8 WRN ERLIN2 PROSC ASH2L LSM1 BAG4 DDHD2 PPAPDC1B WHSC1L1 FGFR1 TACC1 PLEKHA2 TM2D2 GOLGA7 AGPAT6 ANK1 KAT6A AP3M2 PLAT VDAC3 SLC20A2 RNF170 FNTA HGSNAT PRKDC MCM4 UBE2V2 ST18 RB1CC1 ATP6V1H TCEA1 LYPLA1 MRPL15 TMEM68 TGS1 RPS20 CHCHD7 IMPAD1 UBXN2B SDCBP NSMAF TOX RAB2A CHD7 ASPH GGH YTHDF3 ARMC1 C8orf46 VCPIP1 SNHG6 COPS5 CSPP1 ARFGEF1 PREX2 SULF1 TRAM1 TERF1 RPL7 RDH10 STAU2 UBE2W TCEB1 TMEM70 GDAP1 ZFHX4 PEX2 STMN2 HEY1 MRPS28 TPD52 ZNF704 PAG1 PMP2 IMPA1 ZFAND1 RALYL LRRCC1 C8orf59 CA2 WWP1 RMDN1 CPNE3 MMP16 NBN DECR1 TMEM55A OTUD6B RUNX1T1 FAM92A1 TMEM67 PDP1 KIAA1429 DPY19L4 INTS8 UQCRB PTDSS1 CPQ MTDH LAPTM4B MATN2 RPL30 HRSP12 VPS13B COX6C POLR2K SPAG1 RNF19A ANKRD46 PABPC1 YWHAZ ZNF706 NCALD RRM2B UBR5 KLF10 AZIN1 ATP6V1C1 BAALC SLC25A32 DCAF13 RIMS2 OXR1 EIF3E EMC2 NUDCD1 ENY2 SYBU CSMD3 TRPS1 EIF3H UTP23 RAD21 SAMD12 ENPP2 TAF2 MRPL13 DERL1 ZHX1 ATAD2 FBXO32 NDUFB9 SQLE KIAA0196 FAM84B FAM49B ASAP1 PHF20L1 NDRG1 ST3GAL1 KHDRBS3 CHRAC1 PTK2 EEF1D KIAA1875 VPS28 RPL8 ZNF517 ZNF250 ZNF252P C8orf33 KANK1 SMARCA2 RFX3 GLIS3 CDC37L1 AK3 KIAA2026 RANBP6 IL33 GLDC KDM4C PTPRD MPDZ ZDHHC21 SNAPC3 PSIP1 RPS6 SLC24A2 MLLT3 FOCAD KLHL9 MTAP IFT74 MOB3B C9orf72 ACO1 DDX58 NDUFB6 APTX DNAJA1 SMU1 B4GALT1 BAG1 CHMP5 NFX1 NOL6 UBE2R2 UBAP2 DCAF12 UBAP1 KIAA1161 DCTN3 GALT VCP PIGO STOML2 UNC13B TLN1 CREB3 GBA2 GLIPR2 GNE RNF38 ZCCHC7 GRHPR POLR1E TRMT10Baggregating data to list ... 
## Initializing model ... 
## ...Done!AMPLIFICATION PROBABILITY: 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0DELETION PROBABILITY: 0.78925 0.54 0.58225 0.91025 0.56925 0.78575 0.77975 0.511 0.7605 0.78425 0.43225 0.2225 0.63175 0.863 0.21 0.6785 0.9985 0.996 0.7425 0.81075 0.693 0.663 0.36175 0.44225 0.7585 0.9995 0.74875 0.90175 0.6385 0.839 0.64225 0.661 0.8685 0.76375 0.60675 0.40175 0.9715 0.9355 0.35425 0.98175 0.16875 0.27525[1] "CNV SNPS:"
##   [1] "MUT"        "MCM3"       "PAQR8"      "EFHC1"      "TRAM2"     
##   [6] "TMEM14A"    "GSTA4"      "ICK"        "FBXO9"      "ELOVL5"    
##  [11] "LRRC1"      "DST"        "ZNF451"     "PTP4A1"     "PHF3"      
##  [16] "LMBRD1"     "FAM135A"    "RIMS1"      "MTO1"       "SLC17A5"   
##  [21] "COX7A2"     "TMEM30A"    "SENP6"      "MYO6"       "PHIP"      
##  [26] "HMGN3"      "SH3BGRL2"   "BCKDHB"     "IBTK"       "PGM3"      
##  [31] "MRAP2"      "SNX14"      "SYNCRIP"    "SNHG5"      "ZNF292"    
##  [36] "RARS2"      "ORC3"       "CNR1"       "RNGTT"      "PNRC1"     
##  [41] "SRSF12"     "PM20D2"     "RRAGD"      "MDN1"       "CASP8AP2"  
##  [46] "BACH2"      "MAP3K7"     "MANEA"      "FUT9"       "UFL1"      
##  [51] "NDUFAF4"    "POU3F2"     "PNISR"      "USP45"      "CCNC"      
##  [56] "ASCC3"      "GRIK2"      "BVES"       "ATG5"       "QRSL1"     
##  [61] "C6orf203"   "PDSS2"      "SOBP"       "SEC63"      "OSTM1"     
##  [66] "SNX3"       "FOXO3"      "CD164"      "FIG4"       "CDC40"     
##  [71] "CDK19"      "AMD1"       "GTF3C6"     "RPF2"       "REV3L"     
##  [76] "TRAF3IP2"   "FYN"        "TUBE1"      "FAM229B"    "LAMA4"     
##  [81] "MARCKS"     "HDAC2"      "NT5DC1"     "TSPYL4"     "TSPYL1"    
##  [86] "RWDD1"      "SLC35F1"    "CEP85L"     "ASF1A"      "GJA1"      
##  [91] "HSF2"       "SERINC1"    "PKIB"       "FABP7"      "HDDC2"     
##  [96] "NCOA7"      "HINT3"      "TRMT11"     "ECHDC1"     "LAMA2"     
## [101] "EPB41L2"    "AKAP7"      "MED23"      "STX7"       "SLC18B1"   
## [106] "RPS12"      "SGK1"       "HBS1L"      "AHI1"       "PDE7B"     
## [111] "BCLAF1"     "IFNGR1"     "HECA"       "VTA1"       "AIG1"      
## [116] "ADAT2"      "PEX3"       "FUCA2"      "LTV1"       "SF3B5"     
## [121] "UTRN"       "SHPRH"      "SASH1"      "UST"        "TAB2"      
## [126] "PPIL4"      "LATS1"      "NUP43"      "PCMT1"      "PLEKHG1"   
## [131] "MTHFD1L"    "AKAP12"     "RMND1"      "SYNE1"      "MTRF1L"    
## [136] "RGS17"      "IPCEF1"     "ARID1B"     "TMEM242"    "SERAC1"    
## [141] "GTF2H5"     "TULP4"      "TMEM181"    "DYNLT1"     "EZR"       
## [146] "WTAP"       "ACAT2"      "TCP1"       "MRPL18"     "IGF2R"     
## [151] "PARK2"      "QKI"        "SFT2D1"     "MPC1"       "RPS6KA2"   
## [156] "MLLT4"      "THBS2"      "WDR27"      "TBP"        "CSMD1"     
## [161] "MCPH1"      "AGPAT5"     "MFHAS1"     "ERI1"       "TNKS"      
## [166] "MSRA"       "MTMR9"      "NEIL2"      "FDFT1"      "CTSB"      
## [171] "KIAA1456"   "DLC1"       "ZDHHC2"     "CNOT7"      "SLC7A2"    
## [176] "MTUS1"      "PCM1"       "ASAH1"      "PSD3"       "CSGALNACT1"
## [181] "INTS10"     "LPL"        "ATP6V1B2"   "LZTS1"      "SLC39A14"  
## [186] "EGR3"       "CHMP7"      "R3HCC1"     "ENTPD4"     "DOCK5"     
## [191] "PPP2R2A"    "BNIP3L"     "DPYSL2"     "TRIM35"     "PTK2B"     
## [196] "SCARA3"     "CCDC25"     "ELP3"       "FZD3"       "EXTL3"     
## [201] "INTS9"      "HMBOX1"     "KIF13B"     "LEPROTL1"   "DCTN6"     
## [206] "RBPMS"      "UBXN8"      "WRN"        "ERLIN2"     "PROSC"     
## [211] "ASH2L"      "LSM1"       "BAG4"       "DDHD2"      "PPAPDC1B"  
## [216] "WHSC1L1"    "FGFR1"      "TACC1"      "PLEKHA2"    "TM2D2"     
## [221] "GOLGA7"     "AGPAT6"     "ANK1"       "KAT6A"      "AP3M2"     
## [226] "PLAT"       "VDAC3"      "SLC20A2"    "RNF170"     "FNTA"      
## [231] "HGSNAT"     "PRKDC"      "MCM4"       "UBE2V2"     "ST18"      
## [236] "RB1CC1"     "ATP6V1H"    "TCEA1"      "LYPLA1"     "MRPL15"    
## [241] "TMEM68"     "TGS1"       "RPS20"      "CHCHD7"     "IMPAD1"    
## [246] "UBXN2B"     "SDCBP"      "NSMAF"      "TOX"        "RAB2A"     
## [251] "CHD7"       "ASPH"       "GGH"        "YTHDF3"     "ARMC1"     
## [256] "C8orf46"    "VCPIP1"     "SNHG6"      "COPS5"      "CSPP1"     
## [261] "ARFGEF1"    "PREX2"      "SULF1"      "TRAM1"      "TERF1"     
## [266] "RPL7"       "RDH10"      "STAU2"      "UBE2W"      "TCEB1"     
## [271] "TMEM70"     "GDAP1"      "ZFHX4"      "PEX2"       "STMN2"     
## [276] "HEY1"       "MRPS28"     "TPD52"      "ZNF704"     "PAG1"      
## [281] "PMP2"       "IMPA1"      "ZFAND1"     "RALYL"      "LRRCC1"    
## [286] "C8orf59"    "CA2"        "WWP1"       "RMDN1"      "CPNE3"     
## [291] "MMP16"      "NBN"        "DECR1"      "TMEM55A"    "OTUD6B"    
## [296] "RUNX1T1"    "FAM92A1"    "TMEM67"     "PDP1"       "KIAA1429"  
## [301] "DPY19L4"    "INTS8"      "UQCRB"      "PTDSS1"     "CPQ"       
## [306] "MTDH"       "LAPTM4B"    "MATN2"      "RPL30"      "HRSP12"    
## [311] "VPS13B"     "COX6C"      "POLR2K"     "SPAG1"      "RNF19A"    
## [316] "ANKRD46"    "PABPC1"     "YWHAZ"      "ZNF706"     "NCALD"     
## [321] "RRM2B"      "UBR5"       "KLF10"      "AZIN1"      "ATP6V1C1"  
## [326] "BAALC"      "SLC25A32"   "DCAF13"     "RIMS2"      "OXR1"      
## [331] "EIF3E"      "EMC2"       "NUDCD1"     "ENY2"       "SYBU"      
## [336] "CSMD3"      "TRPS1"      "EIF3H"      "UTP23"      "RAD21"     
## [341] "SAMD12"     "ENPP2"      "TAF2"       "MRPL13"     "DERL1"     
## [346] "ZHX1"       "ATAD2"      "FBXO32"     "NDUFB9"     "SQLE"      
## [351] "KIAA0196"   "FAM84B"     "FAM49B"     "ASAP1"      "PHF20L1"   
## [356] "NDRG1"      "ST3GAL1"    "KHDRBS3"    "CHRAC1"     "PTK2"      
## [361] "EEF1D"      "KIAA1875"   "VPS28"      "RPL8"       "ZNF517"    
## [366] "ZNF250"     "ZNF252P"    "C8orf33"    "KANK1"      "SMARCA2"   
## [371] "RFX3"       "GLIS3"      "CDC37L1"    "AK3"        "KIAA2026"  
## [376] "RANBP6"     "IL33"       "GLDC"       "KDM4C"      "PTPRD"     
## [381] "MPDZ"       "ZDHHC21"    "SNAPC3"     "PSIP1"      "RPS6"      
## [386] "SLC24A2"    "MLLT3"      "FOCAD"      "KLHL9"      "MTAP"      
## [391] "IFT74"      "MOB3B"      "C9orf72"    "ACO1"       "DDX58"     
## [396] "NDUFB6"     "APTX"       "DNAJA1"     "SMU1"       "B4GALT1"   
## [401] "BAG1"       "CHMP5"      "NFX1"       "NOL6"       "UBE2R2"    
## [406] "UBAP2"      "DCAF12"     "UBAP1"      "KIAA1161"   "DCTN3"     
## [411] "GALT"       "VCP"        "PIGO"       "STOML2"     "UNC13B"    
## [416] "TLN1"       "CREB3"      "GBA2"       "GLIPR2"     "GNE"       
## [421] "RNF38"      "ZCCHC7"     "GRHPR"      "POLR1E"     "TRMT10B"   
##      MGH31_A02 MGH31_A05 MGH31_A06 MGH31_A07 MGH31_A10 MGH31_B02 MGH31_B04
## [1,]   0.78925      0.54   0.58225   0.91025   0.56925   0.78575   0.77975
##      MGH31_B06 MGH31_B09 MGH31_C02 MGH31_C04 MGH31_C05 MGH31_C06 MGH31_C10
## [1,]     0.511    0.7605   0.78425   0.43225    0.2225   0.63175     0.863
##      MGH31_D03 MGH31_D07 MGH31_D10 MGH31_D12 MGH31_E01 MGH31_E02 MGH31_E03
## [1,]      0.21    0.6785    0.9985     0.996    0.7425   0.81075     0.693
##      MGH31_E05 MGH31_E08 MGH31_E11 MGH31_E12 MGH31_F02 MGH31_F03 MGH31_F04
## [1,]     0.663   0.36175   0.44225    0.7585    0.9995   0.74875   0.90175
##      MGH31_F06 MGH31_F08 MGH31_F11 MGH31_G01 MGH31_G05 MGH31_G08 MGH31_G09
## [1,]    0.6385     0.839   0.64225     0.661    0.8685   0.76375   0.60675
##      MGH31_H02 MGH31_H04 MGH31_H05 MGH31_H09 MGH31_H10 MGH31_H11 MGH31_H12
## [1,]   0.40175    0.9715    0.9355   0.35425   0.98175   0.16875   0.27525
## [1] "GROUP1:"
##  [1] "MGH31_A02" "MGH31_A07" "MGH31_B02" "MGH31_B04" "MGH31_B09"
##  [6] "MGH31_C02" "MGH31_C10" "MGH31_D10" "MGH31_D12" "MGH31_E02"
## [11] "MGH31_E12" "MGH31_F02" "MGH31_F04" "MGH31_F08" "MGH31_G05"
## [16] "MGH31_G08" "MGH31_H04" "MGH31_H05" "MGH31_H10"
## [1] "GROUP2:"
## [1] "MGH31_C05" "MGH31_D03" "MGH31_H11"
## [1] "Recursion for Group1"
## max vote:0max vote:0[1] "Recursion for Group2"
## max vote:0max vote:0[1] "Recursion for Group2"
## [1] "Recursion for Group2"
## [1] "Recursion for Group2"
## [1] "Recursion for Group2"
## max vote:0max vote:0
```

```
## NULL
```

```r
## double check what CNVs were identified
bgf <- get('bound.genes.final', slot(hb, '.xData'))
genes <- get('genes', slot(hb, '.xData'))
range(genes[unlist(bgf)])
```

```
## GRanges object with 11 ranges and 0 metadata columns:
##        seqnames                ranges strand
##           <Rle>             <IRanges>  <Rle>
##    [1]    chr20 [  325401,  64287821]      *
##    [2]     chr6 [49430360, 170572870]      *
##    [3]    chr11 [  167784,   2397419]      *
##    [4]     chr7 [  816615, 158956747]      *
##    [5]     chr9 [  470291, 138124624]      *
##    [6]    chr14 [20256558, 105480170]      *
##    [7]    chr10 [  274190, 133373689]      *
##    [8]    chr19 [57630395,  58555124]      *
##    [9]     chr8 [ 2935353, 145056030]      *
##   [10]    chr13 [20403667, 114305817]      *
##   [11]    chr21 [ 6444869,  15065000]      *
##   -------
##   seqinfo: 78 sequences from an unspecified genome; no seqlengths
```

Retest all identified CNVs on all cells, cluster cells on the final posterior probability, and plot.


```r
hb$retestIdentifiedCnvs()
```

```
## Retesting bound genes ... ERROR NO GENES AFFECTED BY CNVS IDENTIFIED! Run calcGexpCnvBoundaries()?
```

```r
retest <- get('results', slot(hb, '.xData'))[['gene-based']]
results <- do.call(rbind, lapply(retest, function(x) rbind(x[[1]], x[[2]])))
hc1 <- hclust(dist(t(results)), method='ward.D2')
```

```
## Error in t.default(results): argument is not a matrix
```

```r
hb$plotGexpProfile(order=hc1$labels[hc1$order])
```

<img src="figure/unnamed-chunk-6-1.png" title="plot of chunk unnamed-chunk-6" alt="plot of chunk unnamed-chunk-6" width="800px" height="400px" />

```r
head(hb$summarizeResults(geneBased=TRUE))
```

```
## Error in `colnames<-`(`*tmp*`, value = "amp.gexp."): attempt to set 'colnames' on an object with less than two dimensions
```

We can also analyze using allele information. Load the allele data and add them to our HoneyBADGER object.


```r
data(r)
data(cov.sc)

hb$setAlleleMats(r.init=r, n.sc.init=cov.sc, n.cores=1, het.deviance.threshold=0.1)
```

```
## Initializing allele matrices ... 
## creating in-silico bulk ... 
## using 75 cells ... 
## filtering for putative heterozygous snps ... 
## allowing for a 0.1 deviation from the expected 0.5 heterozygous allele fraction ... 
## must have coverage in at least 3 cells ... 
## 5359 heterozygous SNPs identified 
## setting composite minor allele count ... 
## Done setting initial allele matrices!
```

```r
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
hb$setGeneFactors(txdb)
```

```
## Mapping snps to genes ... 
## Done mapping snps to genes!
```

```r
hb$plotAlleleProfile()
```

```
## Error in do.call("grid.arrange", c(plist, ncol = length(plist))): could not find function "grid.arrange"
```

Now run our allele HMM to identify putative regions affected by deletions or LOH.


```r
hb$calcAlleleCnvBoundaries(init=TRUE)
```

```
## Error in matrix(0, nrow(r.maf), ncol(r.maf)): non-numeric matrix extent
```

```r
## double check what CNVs were identified
bsf <- get('bound.snps.final', slot(hb, '.xData'))
snps <- get('snps', slot(hb, '.xData'))
range(snps[unlist(bsf)])
```

```
## [1]  Inf -Inf
```

We can again retest.


```r
retest <- hb$retestIdentifiedCnvs(retestBoundGenes=FALSE, retestBoundSnps=TRUE)
```

```
## Retesting bound snps ... ERROR NO SNPS AFFECTED BY CNVS IDENTIFIED! Run calcAlleleCnvBoundaries()?
```

```r
results <- do.call(rbind, lapply(retest, function(x) x[[1]]))
hc2 <- hclust(dist(t(results)), method='ward.D2')
```

```
## Error in t.default(results): argument is not a matrix
```

```r
hb$plotAlleleProfile(order=hc2$labels[hc2$order])
```

```
## Error in t.default(n.tot[vi, ]): argument is not a matrix
```

```r
head(hb$summarizeResults(geneBased=FALSE, alleleBased=TRUE))
```

```
## Error in `colnames<-`(`*tmp*`, value = "del.loh.allele "): attempt to set 'colnames' on an object with less than two dimensions
```

We can also combine both expression and allele information


```r
hb$retestIdentifiedCnvs(retestBoundGenes=TRUE, retestBoundSnps=TRUE)
```

```
## Retesting bound genes ... ERROR NO GENES AFFECTED BY CNVS IDENTIFIED! Run calcGexpCnvBoundaries()? Retesting bound snps ... ERROR NO SNPS AFFECTED BY CNVS IDENTIFIED! Run calcAlleleCnvBoundaries()? Retesting bound snps and genes using joint modelERROR NO GENES AFFECTED BY CNVS IDENTIFIED! Run calcGexpCnvBoundaries()?
```

```
## NULL
```

```r
results <- get('results', slot(hb, '.xData'))
head(results)
```

```
## list()
```

```r
#hc3 <- hclust(dist(t(results)), method='ward.D2')

hb$plotGexpProfile(order=hc1$labels[hc1$order])
hb$plotGexpProfile(order=hc2$labels[hc2$order])
```

<img src="figure/unnamed-chunk-10-1.png" title="plot of chunk unnamed-chunk-10" alt="plot of chunk unnamed-chunk-10" width="800px" height="400px" />

```r
#hb$plotGexpProfile(order=hc3$labels[hc3$order])

hb$plotAlleleProfile(order=hc1$labels[hc1$order])
```

```
## Error in t.default(n.tot[vi, ]): argument is not a matrix
```

```r
hb$plotAlleleProfile(order=hc2$labels[hc2$order])
```

```
## Error in t.default(n.tot[vi, ]): argument is not a matrix
```

```r
#hb$plotAlleleProfile(order=hc3$labels[h3c$order])

head(hb$summarizeResults(geneBased=TRUE, alleleBased=TRUE))
```

```
## Error in as.vector(x): no method for coercing this S4 class to a vector
```
