---
title: "Preparing Data"
author: "Jean Fan"
date: '`r Sys.Date()`'
output: html_document
md_document:
  variant: markdown_github
vignette: |
  %\VignetteIndexEntry{Vignette Title} %\VignetteEngine{knitr::rmarkdown} \usepackage[utf8]{inputenc}
---


```{r, include = FALSE}
library(knitr)
opts_chunk$set(
    warning = FALSE,
    message = FALSE,
    fig.path = 'figure/',
    cache.path = 'cache/',
    cache = TRUE,
    out.width='800px',
    out.height='400px',
    dpi=100
)
```

Starting with aligned `.bam` files, there are a number of ways to prepare your data for `HoneyBADGER`. 

## Allele Data

For the allele-based model, you will need a set of heterozygous SNP positions. This can be ideally obtained from previous WES data from the same sample or estimated from the [ExAC database](http://exac.broadinstitute.org/).

```{r}
## Download ExAC vcf or use your own vcf
vcfFile <- "data-raw/ExAC.r0.3.sites.vep.vcf.gz"
## For testing purposes, restrict to set of SNPs on region on chromosome 3
require(GenomicRanges)
testRanges <- GRanges('1', IRanges(start=1e5, width=1e6))
require(VariantAnnotation)
param <- ScanVcfParam(which=testRanges)
vcf <- readVcf(vcfFile, "hg19", param=param)
## limit to common snps by MAF
info <- info(vcf)
maf <- info[, 'AF'] # AF is Integer allele frequency for each Alt allele
maft <- 0.1
vi <- sapply(maf, function(x) any(x > maft))
print(table(vi))
snps <- rowData(vcf)
snps <- snps[vi,]
## get rid of non single nucleotide changes
vi <- width(snps@elementMetadata$REF) == 1
snps <- snps[vi,]
## look at final SNPs
print(snps)
```

Now we can get the number of reads corresponding to each SNP site for each cell using their `.bam` files. 

```{r}
library(HoneyBADGER)
path <- "data-raw/"
files <- list.files(path = path)
bamFiles <- files[grepl('.bam$', files)]
bamFiles <- paste0(path, bamFiles)
indexFiles <- files[grepl('.bai$', files)]
indexFiles <- paste0(path, indexFiles)
results <- getSnpMats(snps, bamFiles, indexFiles)
print(names(results))
```

Now we have a matrix of SNP coverage as well as reference and allele count for use in our `HoneyBADGER` allele model. 

```{r}
print(head(results$refCount))
print(head(results$altCount))
print(head(results$cov))
```

## Gene expression data

For gene expression data, we recommend quantification by TPM or FPM transformed to log scale. The same processing pipeline and transformation is highly recommended for the normal reference. Normal references can be ideally obtained from matched normal but can also be estimated using [GTeX](https://www.gtexportal.org/home/). 
